In [1]:
help("keywords")


Here is a list of the Python keywords.  Enter any keyword to get more help.

False               break               for                 not
None                class               from                or
True                continue            global              pass
__peg_parser__      def                 if                  raise
and                 del                 import              return
as                  elif                in                  try
assert              else                is                  while
async               except              lambda              with
await               finally             nonlocal            yield



In [2]:
import math

#### Given a number N>=1, when can we say it is a prime number ?

    When no of factors are == 2
    
    
    
    Number 1 neither prime not composit 
    2 is prime 
    3 is prime 
    4 is composite. ... 
    

## if prime ? 

In [3]:
N = 1
def isprime(N):
    if N == 1:
        return "1 is neither prime nor composite"
    f = 0
    for i in range(1,N+1):
        if N%i == 0:
            f += 1
    return f  == 2
    
isprime(1)

'1 is neither prime nor composite'

In [4]:
isprime(4)

False

In [5]:
isprime(3)

True

In [6]:
N = 1
def isprime(N):
    if N == 1:
        return "1 is neither prime nor composite"
    for i in range(2,N):
        if N%i == 0:
            return False
    return True
    
    
isprime(97)

True

    approximately a computer can handle 10^8 iterations in 1 seconds .
    
    if we want to find if ( 10^9 + 7 ) a prime number ,
    this will take 10 seconds approximately with O(n) algorithm 
    
    if we want to find if ( 10^18) a prime number ,
    above algorithm O(n) will take 10 billion seconds =~ 317 years .
    
    
    so we need to optimize the time taken to run : 

    N = 60

    factors of 60 
    -------------
    1  * 60     i < N
    2  * 30
    3  * 20
    4  * 15
    5  * 12
    6  * 10
    --------factors are repeting after this point (factos appear in pairs) 
    10 * 6      i > N/i
    12 * 5
    15 * 4
    20 * 3
    30 * 2
    60 * 1      
    -------------

    24:
    
    1    24     a < b
    2    12
    3    8
    4    6
    -------------
    6    4      a > b
    8    3
    12   2
    24   1
    
    
    a > b
    a*b = N
    a = N/b
    
    a > b
    N/b > b 
    N > b**2
    
    sq(N) > b
    
    

    instead of looping from 1 to N
    we can only loop till square root of N as after that we wont find any new factors.

In [7]:
import math

In [8]:
N = 1

def isprime(N):
    if N == 1:
        return "1 is neither prime nor composite"
    factor_count = 0
    for i in range(1,(int(math.sqrt(N))+1)):
        if N%i == 0:
            factor_count += 2
    return factor_count == 2
    
    
isprime(97),isprime(98),isprime(13)

(True, False, True)

    # from above code, we will be able to calculate if 10**18 is prime or not in just 10 seconds 

    (math.sqrt(10**18))/10**8 = 10

In [9]:
# if we dont want to use math library or pythonic way (N)**0.5:

    i < sqrt(N) +1 
    i^2 < (sqrt(N)+1)^2
    i^2 < N+1

In [10]:
N = 1
def isprime(N):
    
    i = 1
    factor_count = 0
    
    while i*i <= N:
        
        if N%i == 0:
            factor_count += 2
        i += 1
        
    return factor_count == 2
    
    
isprime(97),isprime(98),isprime(13),isprime(3)

(True, False, True, True)

In [11]:
N = 1
def isprime(N):
    
    i = 2
    
    while i*i <= N:
        
        if N%i == 0:
            return False
        i += 1
        
    return True
    
    
isprime(97),isprime(98),isprime(13),isprime(3),isprime(4),isprime(25)

(True, False, True, True, False, False)

## given a number N find floor largest number  <= sqrt(N)
## ignore the factorial part (round down)
    
        sqrt(49) = 7
        sqrt(24) = 4
        sqrt(25) = 5
        sqrt(26) = 5

In [12]:
# O(sqrt(N))

In [13]:
def returnPerfectSQ_o_sq(N):
    i = 1
    while i*i <= N:
        i += 1
    return i-1
        
returnPerfectSQ_o_sq(25),returnPerfectSQ_o_sq(26),returnPerfectSQ_o_sq(23)

(5, 5, 4)

In [14]:
## using binary search (O(log(n)))

In [15]:
def returnPerfectSQ(N):
    high = N
    low = 1
    
    
    while low <= high:  
        mid = ( high + low ) // 2 
        
        
        if mid * mid > N:
            high = mid - 1
            
            
        elif mid * mid < N:
            low = mid + 1
        
        else:
            return mid
            
        
    return high
            

In [16]:
returnPerfectSQ(26),returnPerfectSQ(25),returnPerfectSQ(24)

(5, 5, 4)

In [17]:
returnPerfectSQ(10**18) # takes 0.6 micro seconds 

1000000000

### Square root of a number

#### Given a number A. Return square root of the number if it is perfect square otherwise return -1.



    Problem Constraints

    1 <= A <= 108



    Input Format

    First argument is an integer A.



    Output Format

    Return an integer which is the square root of A if A is perfect square otherwise return -1.



    Example Input

    Input 1:

    A = 4
    Input 2:

    A = 1001


    Example Output

    Output 1:

    2
    Output 2:

    -1


    Example Explanation

    Explanation 1:

    sqrt(4) = 2
    Explanation 2:

    1001 is not a perfect square.

In [18]:
def perfect_(A):

    def perfect_square(n):

        low = 1
        high = n

        while low <= high: 
            mid = (low + high)//2
            if mid * mid > n:
                high = mid - 1 

            elif mid * mid < n:
                low = mid + 1
            else:
                return mid
        return None

    def sqrt_num_if_perfect_sq(n):
        if perfect_square(n) != None:
            return perfect_square(n) 
        return -1

    return sqrt_num_if_perfect_sq(A)      

In [19]:
perfect_(37)

-1

In [20]:
perfect_(36)

6

 ### Is It Perfect?

#### You are given N positive integers. For each given integer A, you have to tell whether it is a perfect number or not. A perfect number is a positive integer which is equal to the sum of its proper positive divisors.



    Problem Constraints

    1 <= N <= 10

    1 <= A <= 106



    Input Format

    The first line of the input contains a single integer N.

    Each of the next N lines contains a single integer A.



    Output Format

    In a separate line, print YES if a given integer is perfect, else print NO.



    Example Input

    Input 1:

     2
     4
     6 
    Input 2:

     1
     3 


    Example Output

    Output 1:

     NO
     YES 
    Output 2:

     NO 


    Example Explanation

    Explanation 1:

     For A = 4, the answer is "NO" as sum of its proper divisors = 1 + 2 = 3, is not equal to 4. 
     For A = 6, the answer is "YES" as sum of its proper divisors = 1 + 2 + 3 = 6, is equal to 6. 
    Explanation 2:

     For A = 3, the answer is "NO" as sum of its proper divisors = 1, is not equal to 3. 

In [21]:
def isperfect(n):
    factors = 0 
    for i in range(1,n):
        if n%i == 0:
            print(i)
            factors += i
    if factors == n:
        return True
    return False
            
    
isperfect(6)

1
2
3


True

In [22]:
def isperfect(n):
    factors = 0 
    i = 1
    while i < n:
        if n%i == 0:
            print(i)
            factors += i
        i+=1
    if factors == n:
        return True
    return False
            
    
isperfect(6)

1
2
3


True

### Armstrong Numbers!

#### You are given an integer N you need to print all the Armstrong Numbers between 1 to N.  If sum of cubes of each digit of the number is equal to the number itself, then the number is called an Armstrong number.  
    For example, 153 = ( 1 * 1 * 1 ) + ( 5 * 5 * 5 ) + ( 3 * 3 * 3 ).

        Problem Constraints
        1 <= N <= 500

        Input Format
        First and only line of input contains an integer N.

        Output Format
        Output all the Armstrong numbers in range [1,N] each in a new line.

        Example Input
        Input 1:
         5
        Input 2:
         200

        Example Output
        Output 1:
        1
        Output 2:
        1
        153

        Example Explanation
        Explanation 1:
        1 is an armstrong number.
        Explanation 2:
        1 and 153 are armstrong number under 200.

In [23]:
def isArmstrong(n):
    total = 0
    original = n
    while n != 0:
        digit = n % 10
        n = n//10
        total += (digit)**(len(str(original)))
    
    if total == original:
        return True
    return False
    
        

In [24]:
isArmstrong(548834)

True

In [25]:
for i in range(100000):
    if isArmstrong(i):
        print(i)
    

0
1
2
3
4
5
6
7
8
9
153
370
371
407
1634
8208
9474
54748
92727
93084


    N = 2048   

    step   value
    0   2048       N
    1   1024       N/2
    2   512        N/4
    3   256        N/8
    4   128        N/2**4
    5   64         N/2**5 
    6   32
    7   16
    8   8
    9   4
    10  2
    11  1
    .
    .
    .
    k             N/2**k  = 1

    N = 2**k

    logN =  k  = numbers of steps we want

## How many digits are there in 2^46 : 

    10 ^ ?  = 2^46


    2^46 = 10^(46/log2(10))

In [26]:
2**46

70368744177664

In [27]:
int(46/math.log2(10))+1

14

In [28]:
10**(46/math.log2(10))

70368744177664.14

In [29]:
3**15

14348907

In [30]:
int(15/math.log(10,3))+1

8

### Create a list of all squares of 1 till 10

In [31]:
sq = []
for i in range(1,11):
    sq.append(i*i)
print(sq)

[1, 4, 9, 16, 25, 36, 49, 64, 81, 100]


In [32]:
[i*i for i in range(1,11)]

[1, 4, 9, 16, 25, 36, 49, 64, 81, 100]

#### What is the time complexity of accessing element at the ith index in an array of size N? Time complexity to access arr[i] ?

    O(1)

### find the sum of array : 

In [33]:
def sum_array(A):
    total =  0
    for i in range(len(A)):
        total += A[i]
    return total
sum_array([1,2,3,4,5,6,7,8,9,10])

55

In [34]:
def sum_array(A):
    total =  0
    for i in A:
        total += i
    return total
sum_array([1,2,3,4,5,6,7,8,9,10])

55

### Given N elements in list ,count number of elements which have atleast 1 element greater than itself present in the list : 

In [35]:
A = [1,3,2,6,8,9,0,-30,-20,5]

def count_greater_elements(A):
    n = len(A)
    count = 0
    for x in A:
        haslarger = False
        for value in A:
            if value > x:
                haslarger = True
        if haslarger:
            count += 1
    return count
count_greater_elements(A)

9

In [36]:
count_greater_elements([8,8,8,1,8,8,8,2])

2

In [37]:
def count_greater_elements(A):
    
    n = len(A)
    maxval = -float("inf")
    
    for i in A:
        if i > maxval:
            maxval = i

    not_max_count = 0
    
    for x in A:
        if x < maxval:
            not_max_count += 1
            
    return not_max_count
            
    
    
count_greater_elements([8,8,8,1,8,8,8,2])

2

### Reverse the given list : 

In [38]:
A = [1,10,3,6,8]
A[::-1]

[8, 6, 3, 10, 1]

In [39]:
def reverse_list(A):
    
    n = len(A)
    reversed_list = []
    for i in reversed(range(len(A))):
        reversed_list.append(A[i])
        
    return reversed_list

reverse_list(A)

[8, 6, 3, 10, 1]

In [40]:
def reverse_list(A):
    
    n = len(A)
    
    for i in range(len(A)//2):
        A[i],A[n-i-1] = A[n-i-1],A[i]    # tuple unpacking 
    return A

reverse_list(A)

[8, 6, 3, 10, 1]

    i   <>  n-1
    i+1 <>  n-i-1
    
    0 <> 5
    1 <> 4
    2 <> 3 stop here. n//2
    3 <> 2 
    

In [41]:
def reverse_list(A):
    
    n = len(A)
    
    for i in range(len(A)//2):
        temp = A[i]
        A[i] = A[n-i-1]
        A[n-i-1] = temp
        
    return A

A = [1,10,3,6,8]

reverse_list(A)

[8, 6, 3, 10, 1]

###  Rotation Game

### Given an integer array A of size N and an integer B, you have to print the same array after rotating it B times towards the right.


    Problem Constraints
    1 <= N <= 106
    1 <= A[i] <=108
    1 <= B <= 109


    Input Format
    There are 2 lines in the input

    Line 1: The first number is the size N of the array A. Then N numbers follow which indicate the elements in the array A.

    Line 2: A single integer B.


    Output Format
    Print array A after rotating it B times towards the right.


    Example Input
    Input 1 :
    4 1 2 3 4
    2


    Example Output
    Output 1 :
    3 4 1 2


    Example Explanation
    Example 1 :

    N = 4, A = [1, 2, 3, 4] and B = 2.

    Rotate towards the right 2 times - [1, 2, 3, 4] => [4, 1, 2, 3] => [3, 4, 1, 2]

    Final array = [3, 4, 1, 2]


In [42]:
A = list(map(int,input().split()))

rotations = int(input())
n = A.pop(0)
print("length of array: ",n)
print("Array:",A)

rotations = rotations % n

for i in range(rotations):
    A.insert(0,A.pop())
    
A    


5 1 2 3 4 5
2
length of array:  5
Array: [1, 2, 3, 4, 5]


[4, 5, 1, 2, 3]

### Good Pair

### Given an array A and an integer B. A pair(i, j) in the array is a good pair if i != j and (A[i] + A[j] == B). Check if any good pair exist or not.

    Problem Constraints
    1 <= A.size() <= 104
    1 <= A[i] <= 109
    1 <= B <= 109

    Input Format
    First argument is an integer array A.
    Second argument is an integer B.

    Output Format
    Return 1 if good pair exist otherwise return 0.

    Example Input
    Input 1:
    A = [1,2,3,4]
    B = 7
    Input 2:
    A = [1,2,4]
    B = 4
    Input 3:
    A = [1,2,2]
    B = 4

    Example Output
    Output 1:
    1
    Output 2:
    0
    Output 3:
    1

    Example Explanation
    Explanation 1:
     (i,j) = (3,4)
    Explanation 2:
    No pair has sum equal to 4.
    Explanation 3:
     (i,j) = (2,3)

In [49]:
# what is asked n question:

def GoodPair(A,k):
    n = len(A)
    for i in range(n):
        for j in range(i,n):
            if i!=j:
                if A[i]+A[j]==k:
                    return 1
    return 0

A = [1,2,3,4]
k = 7
GoodPair(A,k) 

1

In [50]:
# variations : 

In [53]:
A = [1,2,3,4]
k = 7

n = len(A)
c = 0
for i in range(n):
    for j in range(i,n):
        if A[i]+A[j] == k:
            c+=1
print(c)

A = [1,2,3,4,4,3]
k = 7
   # combination pair counts

1


In [54]:
# individual pair counts 

In [55]:
A = [1,2,3,4]
k = 7

def GoodPair(A,k):
    
    n = len(A)
    c = 0
    Aset = set()
    
    for i in A:
        target = k-i
        if target in Aset:
            c+=1
        else:
            Aset.add(i)
    return c
    

A = [1,2,3,4,2,4]
k = 6
GoodPair(A,k)

2

###  Max and Min of an Array

  



In [56]:
A = list(map(int,input().split()))
N = A[0]
maxm = A[1]


for i in range(1,N+1):
    if A[i] > maxm:
        maxm = A[i]

minm = A[1]
for i in range(1,N):
    if A[i]<minm:
        minm = A[i]

print(maxm,minm)


7 9 8 5 4 1 2 3 5
9 1


### Little Ponny and Maximum Element

#### Little Ponny is given an array, A, of N integers. In a particular operation, he can set any element of the array equal to -1. He wants your help in finding out the minimum number of operations required such that the maximum element of the resulting array is B. If it is not possible, then return -1.


    Problem Constraints
    1 <= |A| <= 105
    1 <= A[i] <= 109

    Input Format
    The first argument of input contains an integer array, A.
    The second argument of input contains an integer, B.

    Output Format
    Return an integer representing the answer.

    Example Input
    Input 1:

     A = [2, 4, 3, 1, 5]
     B = 3 
    Input 2:

     A = [1, 4, 2]
     B = 3 

    Example Output
    Output 1:
     2 
    Output 2:
     -1 

    Example Explanation
    Explanation 1:

     We need to remove 4 and 5 to make 3 the biggest element. 
    Explanation 2:

     As 3 doesn't exist in the array, the answer is -1. 

In [57]:
A = [2, 4, 3, 1, 5]
B = 3 
def little_pony_max_element(A,B):
    if B not in A:
        return -1
    number_of_operations = 0
    n = len(A)
    maxele_to_be = B
    for i in A:
        if i > maxele_to_be:
            number_of_operations +=  1
    return number_of_operations
    

A = [2, 4, 3, 1, 5]
B = 3 
little_pony_max_element(A,B)
    

2

In [58]:

A = [1, 4, 2]
B = 3 
little_pony_max_element(A,B)


-1

### Second Largest

### You are given an integer array A. You have to find the second largest element/value in the array or report that no such element exists.

    Problem Constraints
    1 <= |A| <= 105
    0 <= A[i] <= 109

    Input Format
    The first argument is an integer array A.
    Output Format
    Return the second largest element. If no such element exist then return -1.

    Example Input
    Input 1:
     A = [2, 1, 2] 
    Input 2:
     A = [2]
    Example Output
    Output 1:
     2 
    Output 2:
     -1 

    Example Explanation
    Explanation 1:

     First largest element = 2
     Second largest element = 2
     Third largest element = 1
    Explanation 2:

     There is no second largest element in the array.

In [59]:
A = [1,2,3,4,5]
def find_Second_largest(A):
    n = len(A)
    
    max1 = -float("inf")
    max2 = -float("inf")
    
    for i in A:
        if i > max1:
            max2 = max1
            max1 = i
        elif i > max2:
            max2 = i
    if max2 == -float("inf"):
        return -1,max1
    return max2,max1
    
    
A = [9,4,2,7,3]
find_Second_largest(A)


(7, 9)

In [60]:
A = [99]
find_Second_largest(A)


(-1, 99)

### MINIMUM PICKS

#### You are given an array of integers A of size N. Return the difference between the maximum among all even numbers of A and the minimum among all odd numbers in A.

In [61]:
def minimum_picks(A):
    odds = [x for x in A if x%2!=0]
    even = [x for x in A if x%2==0]



    return max(even)-min(odds)

In [62]:
minimum_picks([5,17,100,1])

99

### Separate Odd Even

#### You are given an integer T denoting the number of test cases. For each test case, you are given an integer array A. You have to print the odd and even elements of array A in 2 separate lines. 
    NOTE: Array elements should have the same relative order as in A.



    Problem Constraints
    1 <= T <= 10

    1 <= |A| <= 105

    1 <= A[i] <= 109



    Input Format
    First line of the input contains a single integer T.

    For each test case:

    First line consists of a single integer |A| denoting the length of array.
    Second line consists of |A| space separated integers denoting the elements of array A.


    Output Format
    For each test case:

    First line should contain an array of space separated integers containing all the odd elements of array A
    Second line should contain an array of space separated integers containing all the even elements of array A


    Example Input
    Input 1:

     2 
     5
     1 2 3 4 5
     3
     4 3 2
    Input 2:

     2 
     3
     2 2 2
     2
     1 1


    Example Output
    Output 1:

     1 3 5
     2 4
     3
     4 2
    Output 2:


     2 2 2
     1 1



    Example Explanation
    Explanation 1:

     Test Case 1: 
     Odd elements in the array(in the order they occur) are: 1, 3, 5.
     Even elements in the array(in the order they occur) are: 2, 4.

     Test Case 2 
     Odd elements in the array(in the order they occur) are: 3.
     Even elements in the array(in the order they occur) are: 4, 2.
    Explanation 2:

     Test Case 1: 
     There are no odd elements in the array.
     Even elements in the array(in the order they occur) are: 2, 2, 2.

     Test Case 2 
     Odd elements in the array(in the order they occur) are: 1, 1.
     There are no even elements in the array.

In [66]:
def main():
    
    T = int(input())
    while T>0:
        T = T-1

        L = int(input())
        A = list(map(int,input().split()))
        
        for i in A:
            if i%2!=0:
                print(i,end =" ")
        print()
        for i in A:
            if i%2==0:
                print(i,end =" ")
        print()


    return 0

if __name__ == '__main__':
    main()

1
1
2

2 


### Given an array of n integers , Q queries are given . 
### In each query start and end index given . print sum of start to end index .

In [69]:
def give_query_results(A,Q):

    while Q > 0:
        start = int(input("start:"))
        end = int(input("end:"))
        ans = 0
        for i in range(start,end+1):
            ans += A[i]
        print("sum:",ans)
        Q -= 1
give_query_results([1, 3, 0, -6, 8, 10, 2, 7],2)

start:1
end:3
sum: -3
start:4
end:5
sum: 18


In [70]:
439-360

79

### Given the scores of the last 10 overs of a cricket match:

    288, 312, 330, 349, 360, 383, 394, 406, 436, 439

#### How many runs were scored in the last 5 overs?

    439-360 = 79



### Given the scores of the last 10 overs of a cricket match :

    288, 312, 330, 349, 360, 383, 394, 406, 436, 439

#### How many runs were scored from 42nd to 45th over (both included) ?

    360-288 = 72


In [71]:
# finding difference score between overs :

In [72]:
cumsum = [288, 312, 330, 349, 360, 383, 394, 406, 436, 439]
score_per_over = []
n = len(cumsum)


for i in range(len(cumsum)-1):
    
    score_per_over.append(cumsum[i+1]-cumsum[i])

score_per_over

[24, 18, 19, 11, 23, 11, 12, 30, 3]

### Prefix Sum (Cumulative Sum) : 

In [73]:
def prefix_sum(A):
    prefixsum = [A[0]]
    for i in A[1:]:
        prefixsum.append(i + prefixsum[-1])
    return prefixsum
    

prefix_sum([1, 3, 0, -6, 4, 7, 8, 9])

[1, 4, 4, -2, 2, 9, 17, 26]

In [74]:
prefix_sum([288,24, 18, 19, 11, 23, 11, 12, 30, 3])

[288, 312, 330, 349, 360, 383, 394, 406, 436, 439]

    if s == 0:
    
    sum_till_e = prefix[e]
    
    else:
    if s != 0:
    
    Sum_Between_S_and_E = prefix[e]-prefix[s-1]

In [76]:
def give_query_results(A,Q):
    
    prefixsum = [A[0]]
    for i in A[1:]:
        prefixsum.append(i + prefixsum[-1])
        
    while Q > 0:
        start = int(input("start:"))
        end = int(input("end:"))
        if start == 0:
            print(prefixsum[end])
        else:
            print(prefixsum[end]-prefixsum[start-1])
        
        Q -= 1
give_query_results([1, 3, 0, -6, 8, 10, 2, 7],2)

start:2
end:3
-6
start:4
end:5
18


#### Sum of all elements from index i to j :

    If i!=0: sum = pf[j] - pf[i-1]

    if i==0: sum = pf[j]


### Given an array of size N return True if there is an equilibrium  index in the array . i is an equilibrium index if 
   #### sum of elements before i = sum of elemebts after i 
   
   #### a[0] + .. + a[i-1] = a[i+1]+...+a[n-1]

In [77]:
# not every array has the equilibrium index . 
# like  [1,2,3,4,5].


In [78]:
def equilibrium_index(A):   # TC:O(^2),SC:O(1)
    
    n = len(A)
    eq_index = []
    for i in range(n):
        before_i = 0
        for j in range(0,i):
            before_i += A[j]
        after_i = 0
        for j in range(i+1,n):
            after_i += A[j]
        if before_i == after_i:
            eq_index.append(i)
    return eq_index
            
        
A = [1,2,3,4,5,8,15]
equilibrium_index(A)

[5]

In [79]:
A = [-7,1,5,2,-4,3,0]
equilibrium_index(A)

[3, 6]

In [80]:
def equilibrium_index(A):   # TC:O(N),SC:O(N)
    
    n = len(A)
    
    prefix = [A[0]]
    for x in A[1:]:
        prefix.append(x+prefix[-1])
    
    
    eq_index = []
    
    
    for i in range(n):
        
        before_i = prefix[i-1]
            
        after_i = prefix[n-1]-prefix[i]
        
        if before_i == after_i:
            eq_index.append(i)
    if len(eq_index) == 0:
        return "No_Equilibrium_index"
    return eq_index
            
        
A = [1,2,3,4,5,8,15]
equilibrium_index(A)

[5]

In [81]:
        
A = [-7,1,5,2,-4,3,0]
equilibrium_index(A)

[3, 6]

In [82]:
        
A = [-7,1,5,2]
equilibrium_index(A)

'No_Equilibrium_index'

In [83]:
        
A = [0,0,0,0,0,0]
equilibrium_index(A)

[0, 1, 2, 3, 4, 5]

In [84]:
# TC:O(N),SC:O(1)

In [85]:
def equilibrium_index(A):
    
    n = len(A)
    
    total = sum(A)
    eq_index = []
        
    before_i = 0
    for i in range(n):
        
        after_i = total - A[i] - before_i
        
        if before_i == after_i:
            eq_index.append(i)
            
        before_i += A[i] # update before_i total after calculating after_i total
    
    return eq_index
            
        
A = [1,2,3,4,5,8,15]
equilibrium_index(A)

[5]

In [86]:
        
A = [-7,1,5,2,-4,3,0]
equilibrium_index(A)

[3, 6]

In [87]:
        
A = [-7,1,5,2]
equilibrium_index(A)

[]

In [88]:
        
A = [0,0,0,0,0,0]
equilibrium_index(A)

[0, 1, 2, 3, 4, 5]

### Range Sum Query

#### You are given an integer array A of length N.
#### You are also given a 2D integer array B with dimensions M x 2, where each row denotes a [L, R] query.
#### For each query, you have to find the sum of all elements from L to R indices in A (1 - indexed).
#### More formally, find A[L] + A[L + 1] + A[L + 2] +... + A[R - 1] + A[R] for each query.

    Problem Constraints
    1 <= N, M <= 105
    1 <= A[i] <= 109
    1 <= L <= R <= N

    Input Format
    The first argument is the integer array A.
    The second argument is the 2D integer array B.

    Output Format
    Return an integer array of length M where ith element is the answer for ith query in B.

    Example Input
    Input 1:
    A = [1, 2, 3, 4, 5]
    B = [[1, 4], [2, 3]]
    Input 2:

    A = [2, 2, 2]
    B = [[1, 1], [2, 3]]


    Example Output
    Output 1:
    [10, 5]
    Output 2:

    [2, 4]


    Example Explanation
    Explanation 1:
    The sum of all elements of A[1 ... 4] = 1 + 2 + 3 + 4 = 10.
    The sum of all elements of A[2 ... 3] = 2 + 3 = 5.
    Explanation 2:

    The sum of all elements of A[1 ... 1] = 2 = 2.
    The sum of all elements of A[2 ... 3] = 2 + 2 = 4.

In [89]:
start,end = [1,4]
print(start)
print(end)

1
4


In [90]:
B = [[1, 4], [2, 3]]

In [91]:
for (start,end) in B:
    print(start,end)
    print()

1 4

2 3



In [92]:
def rangeSum(A, B):


    p = [A[0]]
    for i in range(1,len(A)-1+1):
        p.append(A[i]+p[i-1])

    ans = []

    for (start,end) in B:

        if start == 1:
            sum = p[end-1]
        else:
            sum = p[end-1]-p[start-2]
        ans.append(sum)
    return ans

In [93]:
A = [1, 2, 3, 4, 5]
B = [[1, 4], [2, 3]]
rangeSum(A, B)

[10, 5]

### Special Index

#### Given an array, arr[] of size N, the task is to find the count of array indices such that removing an element from these indices makes the sum of even-indexed and odd-indexed array elements equal.

    Problem Constraints
    1 <= n <= 105
    -105 <= A[i] <= 105

    Input Format
    First argument contains an array A of integers of size N

    Output Format
    Return the count of array indices such that removing an element from these indices makes the sum of even-indexed and odd-indexed array elements equal.

    Example Input
    Input 1:
    A=[2, 1, 6, 4]
    Input 2:
    A=[1, 1, 1]

    Example Output
    Output 1:
    1
    Output 2:
    3

    Example Explanation
    Explanation 1:
    Removing arr[1] from the array modifies arr[] to { 2, 6, 4 } such that, arr[0] + arr[2] = arr[1]. 
    Therefore, the required output is 1. 
    Explanation 2:

    Removing arr[0] from the given array modifies arr[] to { 1, 1 } such that arr[0] = arr[1] 
    Removing arr[1] from the given array modifies arr[] to { 1, 1 } such that arr[0] = arr[1] 
    Removing arr[2] from the given array modifies arr[] to { 1, 1 } such that arr[0] = arr[1] 
    Therefore, the required output is 3.

In [94]:
def special_index(A):
    n = len(A)
    
    count = 0
    for x in range(n):

        temp_list = A[:]
        temp_list.pop(x)
               
        odd_index_sum = 0
        even_index_sum = 0
        nt = len(temp_list)
        
        for i in range(nt):
            if i % 2 == 0:
                
                even_index_sum += temp_list[i]
            else:
                odd_index_sum += temp_list[i]
        if even_index_sum == odd_index_sum:
            count += 1
    return count
            
           
A = [2, 1,3 ,2,0,8,7,1,2,6, 4,2,5,8,4,1]
special_index(A)


3

In [95]:

A = [1,2,0,6,8,3,4,0,7]
special_index(A)

0

In [96]:
def special_index(A):
    n = len(A)
    evenprefix = []
    evensum = 0

    for i in range(n):
        if i % 2 == 0:
            evensum += A[i]
            evenprefix.append(evensum)
        else:
            evenprefix.append(evenprefix[-1])

    oddsum = 0
    oddprefix = [0]
    for i in range(1,n):
        if i % 2 != 0:
            oddsum += A[i]
        oddprefix.append(oddsum)


    count = 0

    for i in range(n):
        evenindex_sum = evenprefix[i-1]+oddprefix[n-1]-oddprefix[i]
        oddindex_sum = oddprefix[i-1]+ evenprefix[n-1]-evenprefix[i]

        if evenindex_sum == oddindex_sum:
            count += 1
    return (count)


A = [2, 1,3 ,2,0,8,7,1,2,6, 4,2,5,8,4,1]
 #  0   1 2  3 4 5 6 7 8 9 10 11 12
special_index(A)


3

In [97]:

A = [1,2,0,6,8,3,4,0,7]
special_index(A)

0

###  Pick from both sides!

#### You are given an integer array A of size N. You have to pick B elements in total. Some (possibly 0) elements from left end of array A and some (possibly 0) from the right end of array A to get the maximum sum.  Find and return this maximum possible sum.

    NOTE: Suppose B = 4, and array A contains 10 elements, then
    You can pick the first four elements or can pick the last four elements, or can pick 1 from front and 3 from the back, etc. You need to return the maximum possible sum of elements you can pick.

    Problem Constraints
    1 <= N <= 105
    1 <= B <= N
    -103 <= A[i] <= 103

    Input Format
    First argument is an integer array A.
    Second argument is an integer B.
    Output Format
    Return an integer denoting the maximum possible sum of elements you picked.

    Example Input
    Input 1:
     A = [5, -2, 3 , 1, 2]
     B = 3
    Input 2:
     A = [1, 2]
     B = 1
    Example Output
    Output 1:
     8
    Output 2:
     2

    Example Explanation
    Explanation 1:

     Pick element 5 from front and element (1, 2) from back so we get 5 + 1 + 2 = 8
    Explanation 2:

     Pick element 2 from end as this is the maximum we can get

In [98]:


def pick_from_both_side(A,B):
    n = len(A)
    maxsum = -float("inf")
    for L in range(B):
        R = B-L

        leftsum = 0
        for i in range(L):
            leftsum += A[i]

        rightsum = 0
        for i in range(n-R,n):
            rightsum += A[i]

        totalsum = rightsum + leftsum
        if totalsum > maxsum:
            maxsum = totalsum
            print(A[0:L],A[n-R:n],"sum:",totalsum)
    
    return (maxsum)

    
A = [7,0,1,-3,2,8,10,6 ,5,-1,0]
B = 5
pick_from_both_side(A,B)

[] [10, 6, 5, -1, 0] sum: 20


20

In [99]:
A = [0,1,8,-2,3,4,-8,-4,1,2,-3,0,-1,-8,2]
B = 6
pick_from_both_side(A,B)

[] [2, -3, 0, -1, -8, 2] sum: -8
[0, 1] [0, -1, -8, 2] sum: -6
[0, 1, 8] [-1, -8, 2] sum: 2
[0, 1, 8, -2, 3] [2] sum: 12


12

In [100]:
def pick_from_both_side(A,B):
    
    n = len(A)
    maxsum = -float("inf")
    
    prefixsum = [A[0]]
    for x in A[1:]:
        prefixsum.append(x+prefixsum[-1])
        
    
    for L in range(B):    # L: leftpick and R:rightpick 
        R = B - L
        

        
        if L == 0:
            leftsum = 0
        else:
            leftsum = prefixsum[L-1]
        
        if n - R == 0:
            rightsum = prefixsum[n-1]
        else:
            rightsum = prefixsum[n-1]-prefixsum[n - R -1]
        
        total = leftsum + rightsum
        
        if total > maxsum:
            maxsum = total
    return maxsum
        
    
A = [5, -2, 3 , 1, 2]
B = 3
pick_from_both_side(A,B)
    


8

In [101]:
pick_from_both_side([0,1,8,-2,3,4,-8,-4,1,2,-3,0,-1,-8,2],6)

12

In [102]:
pick_from_both_side([2,7,8,4,4,-2,1,0,2,5,2,3,0],7)

28

###  Time to equality

#### Given an integer array A of size N. In one second, you can increase the value of one element by 1. Find the minimum time in seconds to make all elements of the array equal.

    Problem Constraints
    1 <= N <= 1000000
    1 <= A[i] <= 1000

    Input Format
    First argument is an integer array A.
    Output Format
    Return an integer denoting the minimum time to make all elements equal.

    Example Input
    A = [2, 4, 1, 3, 2]
    Example Output
    8

    Example Explanation
    We can change the array A = [4, 4, 4, 4, 4]. The time required will be 8 seconds.

In [103]:
def time_to_equality(A):
    maxval = max(A)
    equalizedby = 0
    for i in A:
        diff = maxval - i
        equalizedby += diff
    return equalizedby
        
    
    
A = [2,4,1,3,2]
time_to_equality(A)

8

### Product array puzzle

#### Given an array of integers A, find and return the product array of the same size where the ith element of the product array will be equal to the product of all the elements divided by the ith element of the array.

    Note: It is always possible to form the product array with integer (32 bit) values. Solve it without using the division operator.

    Input Format
    The only argument given is the integer array A.
    Output Format
    Return the product array.
    Constraints

    2 <= length of the array <= 1000
    1 <= A[i] <= 10
    For Example

    Input 1:
        A = [1, 2, 3, 4, 5]
    Output 1:
        [120, 60, 40, 30, 24]

    Input 2:
        A = [5, 1, 10, 1]
    Output 2:
        [10, 50, 5, 50]

In [104]:
def Product_array_puzzle(A):
    product = 1
    for i in A:
        product *= i
    
    for i in range(len(A)):
        A[i] = int(product/A[i])
    return(A)
    

Product_array_puzzle([1,2,3,4,5])

[120, 60, 40, 30, 24]

In [105]:
def Product_array_puzzle(A):

    product = 1
    for i in A:
        product *= i
    
    return [int(product/i) for i in A]
    

Product_array_puzzle([1,2,3,4,5])

[120, 60, 40, 30, 24]

### count good pairs : 

#### given a string ( all lower case ) . COunt the number of pairs (i,j) such that i<j and s[i] = "a" and s[j] = "g".
#### given string S count the number of occurances of subsquence "ag"

In [106]:


def count_good_pairs(S):
    
    n = len(S)
    occurances = 0
    for i in range(n):
        if S[i] == "a":
            for j in range(i+1,n):
                if S[j] == "g":
                    occurances += 1
    return occurances
    
    
S = "axgabbxggaxg"
count_good_pairs(S)


8

In [107]:


def count_good_pairs(S):
    
    n = len(S)
    occurances = 0
    
    gcounts = [0]*n
    gc = 0
    for i in reversed(range(n)):
        if S[i] == "g":
            gc+=1
        gcounts[i] = gc
    
    
    print(gcounts)
    for i in range(n):
        if S[i] == "a":
            occurances += gcounts[i]
    
    
    return occurances
    
    
S = "axgabbxggaxg"
count_good_pairs(S)


[4, 4, 4, 3, 3, 3, 3, 3, 2, 1, 1, 1]


8

In [108]:


def count_good_pairs(S):
    
    n = len(S)
    occurances = 0
    gcounts = 0
    for i in reversed(range(n)):
        if S[i] == "g":
            gcounts += 1
        if S[i] == "a":
            occurances += gcounts
    
    
    
    return occurances
    
    
S = "axgabbxggaxg"
count_good_pairs(S)


8

### given an array A of size n, count the number of leaders in the array : 
#### learder are any number which is strictly greater than all elements on its right .  

In [109]:
A = [1,7,8,-6,0,3,-7,-10]
#        l      l  l  l

In [110]:
def learders_on_the_way(A):
    n = len(A)
    number_of_leaders = 0
    for i in range(n):
        hasleader = True
        for j in range(i+1,n):
            if A[j] >= A[i]:
                hasleader = False
                break
        if hasleader :
            number_of_leaders += 1
    return number_of_leaders
    
A = [1,9,7,8,8,-6,0,3,-7,-10]
learders_on_the_way(A)


5

In [111]:
def learders_on_the_way(A):
    n = len(A)
    number_of_leaders = 0
    for i in range(n):
        hasleader = False
        for j in range(i+1,n):
            if A[j] >= A[i]:
                hasleader = True
                
        if hasleader :
            number_of_leaders += 1
    return number_of_leaders
    
A = [1,9,7,8,8,-6,0,3,-7,-10]
learders_on_the_way(A)


5

In [112]:

def learders_on_the_way(A):
    n = len(A)
    suffixmax = []

    maxxv = A[-1]
    for i in reversed(range(n)):
        maxxv = max(maxxv,A[i])
        suffixmax.append(maxxv)

    print(suffixmax)
    return len(set(suffixmax))

A = [1,9,7,8,8,-6,0,3,-7,-10]
learders_on_the_way(A)

[-10, -7, 3, 3, 3, 8, 8, 8, 9, 9]


5

In [113]:

def learders_on_the_way(A):
    n = len(A)
    maxval = -float("inf")
    count = 0
    for i in reversed(range(n)):
        if A[i] > maxval:
            count += 1
            maxval = A[i]
    return count
            
    
    

A = [1,9,7,8,8,-6,0,3,-7,-10]
learders_on_the_way(A)

5

In [114]:

A = [0]
learders_on_the_way(A)

1

In [115]:

A = []
learders_on_the_way(A)

0

#### Given an integer array A containing N distinct integers, you have to find all the leaders in array A. An element is a leader if it is strictly greater than all the elements to its right side.
    NOTE:The rightmost element is always a leader.
    
    



In [116]:
def leaders_on_the_way(A):
    n = len(A)
    ans = []
    maxnum = -float('inf')
    for i in range(n-1,-1,-1):
        if  A[i] > maxnum :
            ans.append(A[i])
            maxnum = A[i]
    return ans[::-1]

A = [1,9,7,8,8,-6,0,3,-7,-10]
leaders_on_the_way(A)

[9, 8, 3, -7, -10]

### Closest MinMax : 
#### Given an array of size n , find the length of the smallest subarray which contains both min and max of the array .

In [117]:

def closest_min_max(A):
    n = len(A)
    Amin = min(A)
    Amax = max(A)
    minlen = float("inf")
    for s in range(n):
        for e in range(s,n):
            if Amin in A[s:e+1] and Amax in A[s:e+1]:
                length = e-s+1
                
                if length < minlen:
                    minlen = length
                    print(A[s:e+1])
                
    return minlen

A = [0,1,5,-2,3,2,5,5,4,-2,3,0,5]
closest_min_max(A)

[0, 1, 5, -2]
[1, 5, -2]
[5, -2]


2

In [118]:

A = [1, 6, 4, 2, 7, 7, 5, 3, 1, 1, 5]
closest_min_max(A)

[1, 6, 4, 2, 7]
[7, 5, 3, 1]


4

In [119]:
closest_min_max([8, 8, 8, 8, 8, 8])

[8]


1

In [120]:

def closest_min_max(A):
    n = len(A)
    
    Amin = float("inf")
    for i in A:
        if i < Amin:
            Amin = i
            
    Amax = -float("inf")
    for i in A:
        if i > Amax:
            Amax = i
            
    minlen = float("inf")
    
    for s in range(n):
        for e in range(s,n):
            
            hasmin = False
            hasmax = False
            
            
            for i in range(s,e+1):
                if A[i] == Amin:
                    hasmin = True
                if A[i] == Amax:
                    hasmax = True
            if hasmin and hasmax:
                length  = e-s+1
                if length < minlen:
                    minlen = length
                    print(A[s:e+1])
                
    return minlen

A = [0,1,5,-2,3,2,5,5,4,-2,3,0,5]
closest_min_max(A)

[0, 1, 5, -2]
[1, 5, -2]
[5, -2]


2

In [121]:

def closest_min_max(A):
    n = len(A)
    
    Amin = float("inf")
    for i in A:
        if i < Amin:
            Amin = i
            
    Amax = -float("inf")
    for i in A:
        if i > Amax:
            Amax = i
            
    minlen = float("inf")
    
    imin = None
    imax = None
    
    for i in range(n):
        
        if A[i] == Amin:
            imin = i
            if imax is not None:
                length = imin - imax + 1
                minlen = min(length,minlen)
        if A[i] == Amax:
            imax = i
            if imin is not None:
                length = imax - imin + 1
                minlen = min(length,minlen)
            
    return minlen

A = [0,1,5,-2,3,2,5,5,4,-2,3,0,5]
closest_min_max(A)

2

### Bulbs

#### A wire connects N light bulbs. Each bulb has a switch associated with it; however, due to faulty wiring, a switch also changes the state of all the bulbs to the right of the current bulb. Given an initial state of all bulbs, find the minimum number of switches you have to press to turn on all the bulbs.  You can press the same switch multiple times.

    Note: 0 represents the bulb is off and 1 represents the bulb is on.
    Problem Constraints
    0 <= N <= 5×105
    0 <= A[i] <= 1
    Input Format
    The first and the only argument contains an integer array A, of size N.
    Output Format
    Return an integer representing the minimum number of switches required.

    Example Input
    Input 1:
     A = [0, 1, 0, 1]
    Input 2:
     A = [1, 1, 1, 1]
    Example Output
    Output 1:
     4
    Output 2:
     0

    Example Explanation
    Explanation 1:
     press switch 0 : [1 0 1 0]
     press switch 1 : [1 1 0 1]
     press switch 2 : [1 1 1 0]
     press switch 3 : [1 1 1 1]
    Explanation 2:
     There is no need to turn any switches as all the bulbs are already on.

In [122]:
A = [0,1,0,1,0,1,1,0]
state = 0
count = 0
n = len(A)
for i in range(n):
    if A[i] == state:
        count += 1
        state = 1- state
        A[i] = state
print(count)

7


### Amazing Subarrays: 

#### You are given a string S, and you have to find all the amazing substrings of S. An amazing Substring is one that starts with a vowel (a, e, i, o, u, A, E, I, O, U).

    Input

    Only argument given is string S.
    Output

    Return a single integer X mod 10003, here X is the number of Amazing Substrings in given the string.
    Constraints

    1 <= length(S) <= 1e6
    S can have special characters
    Example

    Input
    ABEC

    Output
    6

    Explanation
    Amazing substrings of given string are :
    1. A
    2. AB
    3. ABE
    4. ABEC
    5. E
    6. EC
    here number of substrings are 6 and 6 % 10003 = 6.

In [123]:
A = "ABEC"
n = len(A)
x = len(A)
count_of_subarrays = 0
vowels = set({'a', 'e', 'i', 'o', 'u', 'A', 'E', 'I', 'O', 'U'})
for i in range(n):
    if A[i] in vowels:
        count_of_subarrays += x
    x -= 1
print(count_of_subarrays % 10003)

6


In [124]:
A = "ABEC"
n = len(A)
x = len(A)
count_of_subarrays = 0
vowels = set({'a', 'e', 'i', 'o', 'u', 'A', 'E', 'I', 'O', 'U'})
for s in range(n):
    for e in range(s,n):
        if A[s] in vowels:
            print(A[s:e+1])
            count_of_subarrays += 1
print(count_of_subarrays % 10003)

A
AB
ABE
ABEC
E
EC
6


### Even Subarrays

#### You are given an integer array A. Decide whether it is possible to divide the array into one or more subarrays of even length such that the first and last element of all subarrays will be even. Return "YES" if it is possible; otherwise, return "NO" (without quotes).

    Problem Constraints
    1 <= |A|, A[i] <= 106
    Input Format
    The first and the only input argument is an integer array, A.
    Output Format
    Return a string "YES" or "NO" denoting the answer.

    Example Input
    Input 1:
     A = [2, 4, 8, 6]
    Input 2:
     A = [2, 4, 8, 7, 6]
    Example Output
    Output 1:
     "YES"
    Output 2:
     "NO"
    Example Explanation
    Explanation 1:
     We can divide A into [2, 4] and [8, 6].
    Explanation 2:
     There is no way to divide the array into even length subarrays.

In [125]:
A = [2,4,8,7,6]
def even_subarrays(A):
    if len(A)%2 == 0 and A[0]%2 == 0 and A[-1]%2 == 0:
        return "YES"
    else:
        return "NO"
    
A = [2,4,8,7,6]
even_subarrays(A)

'NO'

In [126]:
A = [2,4,8,6]
even_subarrays(A)

'YES'

In [127]:
A = [2,4,8,5]
even_subarrays(A)

'NO'

In [128]:
A = [2,4,5,6]
even_subarrays(A)

'YES'

### Print subarrays of given array of size n :

In [129]:
A = [1,2,3,4,5]

n = len(A)
for s in range(n):
    for e in range(s,n):
        for i in range(s,e+1):
            print(A[i],end = " ")
        print()

1 
1 2 
1 2 3 
1 2 3 4 
1 2 3 4 5 
2 
2 3 
2 3 4 
2 3 4 5 
3 
3 4 
3 4 5 
4 
4 5 
5 


### Find Maxuimum sum Subarray :

In [130]:

def max_subarray_sum(A):
    n = len(A)
    maxsum = -float("inf")
    
    prefixsum = [A[0]]
    for x in A[1:]:
        prefixsum.append(x + prefixsum[-1])
    print(prefixsum)
    
    for s in range(n):
        for e in range(s,n):
            if s == 0:
                subsum = prefixsum[e]
            else:
                subsum = prefixsum[e] - prefixsum[s-1]
            if subsum > maxsum:
                maxsum = subsum
                print(A[s:e+1],subsum)
    return maxsum
    
    
A = [5,-10,100,-20,15,-10,25,30]

max_subarray_sum(A)

[5, -5, 95, 75, 90, 80, 105, 135]
[5] 5
[5, -10, 100] 95
[5, -10, 100, -20, 15, -10, 25] 105
[5, -10, 100, -20, 15, -10, 25, 30] 135
[100, -20, 15, -10, 25, 30] 140


140

In [131]:

def max_subarray_sum(A):
    n = len(A)
    maxsum = -float("inf")
    subarray_maxsum = None
    for s in range(n):
        subtotal = 0
        for e in range(s,n):
            subtotal += A[e]
            
            if subtotal > maxsum:
                maxsum = subtotal
                subarray_maxsum = A[s:e+1]
                print(subarray_maxsum,maxsum)
            
            
    return maxsum
    
A = [5,-10,100,-20,15,-10,25,30]

max_subarray_sum(A)

[5] 5
[5, -10, 100] 95
[5, -10, 100, -20, 15, -10, 25] 105
[5, -10, 100, -20, 15, -10, 25, 30] 135
[100, -20, 15, -10, 25, 30] 140


140

In [132]:
    
A = [-15,8,-2,1,2,10,-5,-7,3,2,1,-1]
max_subarray_sum(A)

[-15] -15
[-15, 8] -7
[-15, 8, -2, 1, 2] -6
[-15, 8, -2, 1, 2, 10] 4
[8] 8
[8, -2, 1, 2] 9
[8, -2, 1, 2, 10] 19


19

In [133]:
# Kaden's Aalgorithm 

In [134]:

def max_subarray_sum(A):
    n = len(A)
    maxsum = -float("inf")
    
    
    current_max = A[0]
    
    for i in range(1,n):
        current_max = max( A[i], current_max + A[i] )
        maxsum = max( maxsum, current_max )
    
    
            
            
    return maxsum
    
A = [5,-10,100,-20,15,-10,25,30]

max_subarray_sum(A)

140

#### Given N array elements how many subarrays of len = 1 are there ?
    
    N


#### Given N array elements how many subarrays of len = 2 are there ?
    
    N-1

#### Given N array elements how many subarrays of len = k are there ?
    
    N-k+1

#### Given N array elements and len = k ? Print last subarray start index and last index
    
    [N-K , N-1]


    e-s+1 = length 
    n-1 - s +1 = k  
    s = n-k
    and 
    e = n-1


### Maximum subarray sum with length k 

In [135]:
def max_sum_sub_with_k_len(A,k):
    
    n = len(A)
    
    maxsum = -float("inf")
    
    for s in range(n-k+1):
        e = k + s -1
        subsum = 0
        for i in range(s,e+1):
            subsum += A[i]
            if subsum > maxsum:
                maxsum = subsum
                print(A[s:e+1],subsum)
    return maxsum
    
    
max_sum_sub_with_k_len([5,-10,100,-20,15,-10,25,30],5)

[5, -10, 100, -20, 15] 5
[5, -10, 100, -20, 15] 95
[100, -20, 15, -10, 25] 100
[100, -20, 15, -10, 25] 110


110

In [136]:
def max_sum_sub_with_k_len(A,k):
    
    n = len(A)
    
    maxsum = -float("inf")
    
    prefixsum = [A[0]]
    for x in A[1:]:
        prefixsum.append(x+prefixsum[-1])
        
    
    
    for s in range(n-k+1):
        e = k + s -1
        subsum = 0
        
        if s == 0:
            subsum = prefixsum[e]
        else:
            subsum = prefixsum[e]- prefixsum[s-1]
            
        if subsum > maxsum:
            maxsum = subsum
            print(A[s:e+1],subsum)
    return maxsum
    
    
max_sum_sub_with_k_len([5,-10,100,-20,15,-10,25,30],5)

[5, -10, 100, -20, 15] 90
[100, -20, 15, -10, 25] 110


110

In [137]:
max_sum_sub_with_k_len([5,-10,100,-20,15,-10,25,30],5)

[5, -10, 100, -20, 15] 90
[100, -20, 15, -10, 25] 110


110

In [138]:
def max_sum_sub_with_k_len(A,k):
    
    n = len(A)
    
    ksum = sum(A[:k])
    
       
    
    maxsum = -float("inf")
    
    for s in range(1,n-k+1):
        e = s+k-1
        
        ksum = ksum - A[s-1]  + A[e]
        
        
        if ksum > maxsum:
            maxsum = ksum
            
    return maxsum
    
    
max_sum_sub_with_k_len([5,-10,100,-20,15,-10,25,30],5)

110

### . Subarray with least average

### Given an array of size N, find the subarray of size K with the least average.

    Problem Constraints
    1<=k<=N<=1e5
    -1e5<=A[i]<=1e5
    Input Format
    First argument contains an array A of integers of size N.
    Second argument contains integer k.
    Output Format
    Return the index of the first element of the subarray of size k that has least average.
    Array indexing starts from 0.

    Example Input
    Input 1:
    A=[3, 7, 90, 20, 10, 50, 40]
    B=3
    Input 2:
    A=[3, 7, 5, 20, -10, 0, 12]
    B=2
    Example Output
    Output 1:
    3
    Output 2:
    4

    Example Explanation
    Explanation 1:
    Subarray between indexes 3 and 5
    The subarray {20, 10, 50} has the least average 
    among all subarrays of size 3.
    Explanation 2:

     Subarray between [4, 5] has minimum average

In [139]:
def subarray_with_least_avg_k_size(A,k):
    
    n = len(A)
    minavg = float("inf")
    
    ksum = sum(A[:k])
    min_sub_array_index = None
    
    for i in range(1,n-k+1):
        j = k + i -1
        ksum = ksum - A[i-1] + A[j]
        
        kavg = ksum/(j-i+1)
        
        if kavg < minavg:
            minavg = kavg
            min_sub_array_index = i
    return min_sub_array_index
            
    
    

    
A = [3, 7, 90, 20, 10, 50, 40]
k = 3

subarray_with_least_avg_k_size(A,k)
    

3

In [140]:
A=[3, 7, 5, 20, -10, 0, 12]
B=2
subarray_with_least_avg_k_size(A,B)


4

###  Maximum Subarray Easy : 


#### You are given an integer array C of size A. Now you need to find a subarray (contiguous elements) so that the sum of contiguous elements is maximum. But the sum must not exceed B.

    Problem Constraints
    1 <= A <= 103
    1 <= B <= 109
    1 <= C[i] <= 106

    Input Format
    The first argument is the integer A.
    The second argument is the integer B.
    The third argument is the integer array C.

    Output Format
    Return a single integer which denotes the maximum sum.

    Example Input
    Input 1:
    A = 5
    B = 12
    C = [2, 1, 3, 4, 5]
    Input 2:
    A = 3
    B = 1
    C = [2, 2, 2]

    Example Output
    Output 1:
    12
    Output 2:
    0

    Example Explanation
    Explanation 1:
    We can select {3,4,5} which sums up to 12 which is the maximum possible sum.
    Explanation 2:

    All elements are greater than B, which means we cannot select any subarray.
    Hence, the answer is 0.

In [141]:

def max_sub_easy(A,k):
    n = len(A)
    maxum = -float("inf")
    for s in range(n):
        subsum = 0
        for e in range(s,n):
            subsum += A[e]
            if subsum <= k:
                maxum = max(maxum, subsum)
            else:
                break
    if maxum == -float("inf"):
        return 0
    return maxum
                
n = 5
k = 12
A = [2, 1, 3, 4, 5]
max_sub_easy(A,k)
 

12

In [142]:
A = 3
B = 1
C = [2, 2, 2]
max_sub_easy(C,B)

0

###  Counting Subarrays!

#### Given an array A of N non-negative numbers and you are also given non-negative number B. You need to find the number of subarrays in A having sum less than B. We may assume that there is no overflow.

    Problem Constraints
    1 <= N <= 104
    1 <= A[i] <= 100
    1 <= B <= 108

    Input Format
    First argument is an integer array A.
    Second argument is an integer B.
    Output Format
    Return an integer denoting the number of subarrays in A having sum less than B.

    Example Input
    Input 1:
     A = [2, 5, 6]
     B = 10
    Input 2:
     A = [1, 11, 2, 3, 15]
     B = 10
    Example Output
    Output 1:
     4
    Output 2:
     4

    Example Explanation
    Explanation 1:
     The subarrays with sum less than B are {2}, {5}, {6} and {2, 5},
    Explanation 2:
     The subarrays with sum less than B are {1}, {2}, {3} and {2, 3}

In [143]:
A = [1, 11, 2, 3, 15]
B = 10


count = 0
n = len(A)
for s in range(n):
    subsum = 0
    for e in range(s,n):
        subsum += A[e]
        if subsum < B:
            count += 1
print(count)
        

4


### Good Subarrays Easy

#### Given an array of integers A, a subarray of an array is said to be good if it fulfills any one of the criteria: 1. Length of the subarray is be even, and the sum of all the elements of the subarray must be less than B. 2. Length of the subarray is be odd, and the sum of all the elements of the subarray must be greater than B. Your task is to find the count of good subarrays in A.

    Problem Constraints
    1 <= len(A) <= 103
    1 <= A[i] <= 103
    1 <= B <= 107
    Input Format
    The first argument given is the integer array A.
    The second argument given is an integer B.
    Output Format
    Return the count of good subarrays in A.
    Example Input
    Input 1:
    A = [1, 2, 3, 4, 5]
    B = 4
    Input 2:
    A = [13, 16, 16, 15, 9, 16, 2, 7, 6, 17, 3, 9]
    B = 65
    Example Output
    Output 1:
    6
    Output 2:
    36
    Example Explanation
    Explanation 1:
    Even length good subarrays = {1, 2}
    Odd length good subarrays = {1, 2, 3}, {1, 2, 3, 4, 5}, {2, 3, 4}, {3, 4, 5}, {5} 

In [144]:
def Good_Subarrays_Easy(A,k):
    n = len(A)
    
    count = 0
    
    for s in range(n):
        subsum = 0
        for e in range(s,n):
            subsum += A[e]
            length = e-s+1
            
            if length % 2 == 0 and subsum < B:
                count += 1
            if length % 2 != 0 and subsum > B:
                count += 1
    return count
    
A = [13, 16, 16, 15, 9, 16, 2, 7, 6, 17, 3, 9]
B = 65 
Good_Subarrays_Easy(A,B)

36

In [145]:
A = [1, 2, 3, 4, 5]
B = 4
Good_Subarrays_Easy(A,B)

6

### Alternating Subarrays Easy : 

### You are given an integer array A of length N comprising of 0's & 1's, and an integer B. You have to tell all the indices of array A that can act as a center of 2 * B + 1 length 0-1 alternating subarray. A 0-1 alternating array is an array containing only 0's & 1's, and having no adjacent 0's or 1's. For e.g. arrays [0, 1, 0, 1], [1, 0] and [1] are 0-1 alternating, while [1, 1] and [0, 1, 0, 0, 1] are not.

    Problem Constraints
    1 <= N <= 103
    A[i] equals to 0 or 1.
    0 <= B <= (N - 1) / 2

    Input Format
    First argument is an integer array A.
    Second argument is an integer B.

    Output Format
    Return an integer array containing indices(0-based) in sorted order. If no such index exists, return an empty integer array.

    Example Input
    Input 1:
     A = [1, 0, 1, 0, 1]
     B = 1 
    Input 2:
     A = [0, 0, 0, 1, 1, 0, 1]
     B = 0 

    Example Output
    Output 1:
     [1, 2, 3]
    Output 2:
     [0, 1, 2, 3, 4, 5, 6]

    Example Explanation
    Explanation 1:
     Index 1 acts as a centre of alternating sequence: [A0, A1, A2]
     Index 2 acts as a centre of alternating sequence: [A1, A2, A3]
     Index 3 acts as a centre of alternating sequence: [A2, A3, A4] 
    Explanation 2:
     Each index in the array acts as the center of alternating sequences of lengths 1.

In [146]:
A = [0,0,0,1,1,0,1]
B = 0
def Alternating_Subarrays_Easy(A,B):
    n = len(A)
    l = 2*B +1 
    ans = []
    for s in range(n-l+1):
        e = l+s-1

        isalternating = True
        prev_val = None
        for i in range(s,e+1):
            if A[i] == prev_val:
                isalternating = False
                break
            prev_val = A[i]

        if isalternating:
            ans.append((s+e)//2)
    return ans

A = [0,0,0,1,1,0,1]
B = 0
Alternating_Subarrays_Easy(A,B)

[0, 1, 2, 3, 4, 5, 6]

In [147]:
A = [1,0,1,0,1]
B = 1
Alternating_Subarrays_Easy(A,B)

[1, 2, 3]

In [148]:

A = [1,0,1,0,1,0,0,0,1,1,1,0,1,0,1,1,1,1,1,0,0]
B = 1
Alternating_Subarrays_Easy(A,B)

[1, 2, 3, 4, 11, 12, 13]

In [149]:

def Alternating_Subarrays_Easy(A,B):
    n = len(A)
    l = 2*B +1 
    ans = []
    
    if l > n:
        return 0
    
    string = ""
    
    for i in A:
        string += str(i)
    print(string)
    
    for i in range(B,n-B):
        if "00" not in string[i-B:i+B+1] and "11" not in string[i-B:i+B+1]:
            ans.append(i)
    return ans
    
    
A = [0,0,0,1,1,0,1]
B = 0
Alternating_Subarrays_Easy(A,B)

0001101


[0, 1, 2, 3, 4, 5, 6]

In [150]:

A = [1,0,1,0,1]
B = 1
Alternating_Subarrays_Easy(A,B)

10101


[1, 2, 3]

In [151]:

A = [1,0,1,0,1,0,0,0,1,1,1,0,1,0,1,1,1,1,1,0,0]
B = 1
Alternating_Subarrays_Easy(A,B)

101010001110101111100


[1, 2, 3, 4, 11, 12, 13]

### In how many subarrays, the element at index 1 will be present?
    10

### In an array of size N, how many subarrays will contain the element at index i?
    (i+1)*(N-i)

### given array A find the sum of all elements of all subarrays :

In [152]:
# bruteforce 

In [153]:
A = [1,2,3,4,5]
n = len(A)
total = 0
for s in range(n):
    for e in range(s,n):
        for i in range(s,e+1):
            total += A[i]
print(total)

105


In [154]:
# prefix sum technique

In [155]:
A = [1,2,3,4,5]
n = len(A)
prefixsum = [A[0]]
for x in A[1:]:
    prefixsum.append(x + prefixsum[-1])

total = 0

for s in range(n):
    sm = 0
    for e in range(s,n):
        if s == 0:
            sm = prefixsum[e]
        else:
            sm = prefixsum[e]-prefixsum[s-1]
        total += sm
print(total)

105


In [156]:
# carry forward technique

In [157]:
A = [1,2,3,4,5]
n = len(A)
total = 0
for s in range(n):
    subsm = 0
    for e in range(s,n):
        subsm += A[e]
        total += subsm
print(total)

105


In [158]:
# inverting problem

In [159]:
A = [1,2,3,4,5]
n = len(A)
total = 0
for i in range(n):          # all elements in array
    
    total += A[i] * (n-i)*(i+1)   # multiplying each element with number of subarrays 
                                 # that particular element is contained in . 
                                # and adding all the values as total sum 
    
    
print(total)

105


In [160]:
#  [ 1 2 3 4 5 6 7 8 9 10 ]
#    0 1 2 3 4 5 6 7 8 9
#            4 
#            i
#  [   i+1    ]
#           [    n-i     ]



A = [1,2,3,4,5,6,7,8,9,10]
#    0 1 2 3 4 5 6 7 8 9 
n = len(A)
total = 0
for i in range(n):          # all elements in array
    print("element:",A[i]," is contained in ",(n-i)*(i+1)," subarrays", " so total :",A[i] * (n-i)*(i+1))
    total += A[i] * (n-i)*(i+1)   # multiplying each element with number of subarrays 
                                 # that particular element is contained in . 
                                # and adding all the values as total sum 
    
print()  
print("sum of all subarrays: ",total)

element: 1  is contained in  10  subarrays  so total : 10
element: 2  is contained in  18  subarrays  so total : 36
element: 3  is contained in  24  subarrays  so total : 72
element: 4  is contained in  28  subarrays  so total : 112
element: 5  is contained in  30  subarrays  so total : 150
element: 6  is contained in  30  subarrays  so total : 180
element: 7  is contained in  28  subarrays  so total : 196
element: 8  is contained in  24  subarrays  so total : 192
element: 9  is contained in  18  subarrays  so total : 162
element: 10  is contained in  10  subarrays  so total : 100

sum of all subarrays:  1210


## 2D arrays : 

In [161]:
M = [ [i*i+j for i in range(5)] for j in range(5)]

In [162]:
M


[[0, 1, 4, 9, 16],
 [1, 2, 5, 10, 17],
 [2, 3, 6, 11, 18],
 [3, 4, 7, 12, 19],
 [4, 5, 8, 13, 20]]

In [163]:
M[0]

[0, 1, 4, 9, 16]

In [164]:
M[0][-1]

16

In [165]:
M[-1][-1]

20

   
       n = 5
       m = 5
       
    n:m    0  1  2   3   4
    
     0   [[0, 1, 4,  9, 16]
     1    [1, 2, 5, 10, 17]
     2    [2, 3, 6, 11, 18]
     3    [3, 4, 7, 12, 19]
     4    [4, 5, 8, 13, 20]]

In [166]:
M[4][4],M[1][-1]

(20, 17)

### What is the bottom right cell in a matrix of size N * M?
    N-1, M-1
    
### What is the top right cell in a matrix of size N * M?
    0, M-1
    
### Given a matrix of size NxM. What will be the index of the bottom left cell?
    N-1,0
   
### Given a matrix of size NxM. What will be the index of the top left cell?
    0, 0
   


In [167]:
n = 5
m = 8


In [168]:
M = [ [0 for i in range(m)] for _ in range(n)]

In [169]:
M

[[0, 0, 0, 0, 0, 0, 0, 0],
 [0, 0, 0, 0, 0, 0, 0, 0],
 [0, 0, 0, 0, 0, 0, 0, 0],
 [0, 0, 0, 0, 0, 0, 0, 0],
 [0, 0, 0, 0, 0, 0, 0, 0]]

In [170]:
x = 1
for i in range(n):
    for j in range(m):
        M[i][j] = x
        x = x+1
M

[[1, 2, 3, 4, 5, 6, 7, 8],
 [9, 10, 11, 12, 13, 14, 15, 16],
 [17, 18, 19, 20, 21, 22, 23, 24],
 [25, 26, 27, 28, 29, 30, 31, 32],
 [33, 34, 35, 36, 37, 38, 39, 40]]

In [171]:
M

[[1, 2, 3, 4, 5, 6, 7, 8],
 [9, 10, 11, 12, 13, 14, 15, 16],
 [17, 18, 19, 20, 21, 22, 23, 24],
 [25, 26, 27, 28, 29, 30, 31, 32],
 [33, 34, 35, 36, 37, 38, 39, 40]]

In [172]:
len(M)

5

In [173]:
len(M[0])

8

In [174]:
M[0:2]

[[1, 2, 3, 4, 5, 6, 7, 8], [9, 10, 11, 12, 13, 14, 15, 16]]

In [175]:
M[2:4]

[[17, 18, 19, 20, 21, 22, 23, 24], [25, 26, 27, 28, 29, 30, 31, 32]]

### Print matrix 

In [176]:
M = [[1, 2, 3, 4, 5, 6, 7, 8],
 [9, 10, 11, 12, 13, 14, 15, 16],
 [17, 18, 19, 20, 21, 22, 23, 24],
 [25, 26, 27, 28, 29, 30, 31, 32],
 [33, 34, 35, 36, 37, 38, 39, 40]]
n = len(M)
m = len(M[0])
(n,m)

(5, 8)

In [177]:
for i in range(n):
    for j in range(m):
        print(M[i][j], end = " ")
    print()

1 2 3 4 5 6 7 8 
9 10 11 12 13 14 15 16 
17 18 19 20 21 22 23 24 
25 26 27 28 29 30 31 32 
33 34 35 36 37 38 39 40 


In [178]:
for row in M:
    for ele in row:
        print(ele, end = " ")
    print()

1 2 3 4 5 6 7 8 
9 10 11 12 13 14 15 16 
17 18 19 20 21 22 23 24 
25 26 27 28 29 30 31 32 
33 34 35 36 37 38 39 40 


## Transpose matrix 

In [179]:
M = [[s+i for i in range(1,9)] for s in range (10,50,10) ]
M

[[11, 12, 13, 14, 15, 16, 17, 18],
 [21, 22, 23, 24, 25, 26, 27, 28],
 [31, 32, 33, 34, 35, 36, 37, 38],
 [41, 42, 43, 44, 45, 46, 47, 48]]

In [180]:
m = len(M[0])
n = len(M)

In [181]:
m,n

(8, 4)

In [182]:
for icol in range(m):
    for irow in range(n):
        print(M[irow][icol], end = " ")
    print()

11 21 31 41 
12 22 32 42 
13 23 33 43 
14 24 34 44 
15 25 35 45 
16 26 36 46 
17 27 37 47 
18 28 38 48 


In [183]:
A = np.array([[11, 12, 13, 14, 15, 16, 17, 18],
 [21, 22, 23, 24, 25, 26, 27, 28],
 [31, 32, 33, 34, 35, 36, 37, 38],
 [41, 42, 43, 44, 45, 46, 47, 48]])
A.T

<IPython.core.display.Javascript object>

array([[11, 21, 31, 41],
       [12, 22, 32, 42],
       [13, 23, 33, 43],
       [14, 24, 34, 44],
       [15, 25, 35, 45],
       [16, 26, 36, 46],
       [17, 27, 37, 47],
       [18, 28, 38, 48]])

## Print sum of each row 

In [184]:

def print_row_sum(M):
    m = len(M[0])
    n = len(M)
    for i in range(n):
        rowsum = 0
        for j in range(m):
            rowsum += M[i][j]
            print(M[i][j],end = " ")
        print(":",rowsum)
        
    

M = [[s+i for i in range(1,9)] for s in range (10,50,10) ]
print_row_sum(M)

11 12 13 14 15 16 17 18 : 116
21 22 23 24 25 26 27 28 : 196
31 32 33 34 35 36 37 38 : 276
41 42 43 44 45 46 47 48 : 356


## print sum of each columns : 

In [185]:

def print_row_sum(M):
    m = len(M[0])
    n = len(M)
    for col in range(m):
        colsum = 0
        for row in range(n):
            colsum += M[row][col]
            print(M[row][col],end = " ")
        print(":",colsum)
        
    

M = [[s+i for i in range(1,9)] for s in range (10,50,10) ]
print_row_sum(M)

11 21 31 41 : 104
12 22 32 42 : 108
13 23 33 43 : 112
14 24 34 44 : 116
15 25 35 45 : 120
16 26 36 46 : 124
17 27 37 47 : 128
18 28 38 48 : 132


## given a sq matrix print diagonals : 

In [186]:
M = [[s+i for i in range(1,6)] for s in range (10,60,10) ]
M

[[11, 12, 13, 14, 15],
 [21, 22, 23, 24, 25],
 [31, 32, 33, 34, 35],
 [41, 42, 43, 44, 45],
 [51, 52, 53, 54, 55]]

In [187]:
# primary diagonals : 

for i in range(len(M)):
    print(M[i][i])

11
22
33
44
55


In [188]:
# secondary diagonal : 

In [189]:
n = len(M)
m = len(M[0])
r = 0
c = m-1
while c >= 0 and r <= n-1:
    print(M[r][c])
    c -= 1
    r += 1

15
24
33
42
51


In [190]:
for r in range(n):
    c = n-r-1
    print(M[r][c])

15
24
33
42
51


In [191]:
# print diagonal : 

In [192]:
row = 4
col = 8

In [193]:
M = [[0]*col for _ in range(row)]
M

[[0, 0, 0, 0, 0, 0, 0, 0],
 [0, 0, 0, 0, 0, 0, 0, 0],
 [0, 0, 0, 0, 0, 0, 0, 0],
 [0, 0, 0, 0, 0, 0, 0, 0]]

In [194]:
for i in range(row):
    for j in range(col):
        M[i][j]= (j+1)**(i+1)

In [195]:
M

[[1, 2, 3, 4, 5, 6, 7, 8],
 [1, 4, 9, 16, 25, 36, 49, 64],
 [1, 8, 27, 64, 125, 216, 343, 512],
 [1, 16, 81, 256, 625, 1296, 2401, 4096]]

In [196]:
M = []
for i in range(row):
    M.append([])
    for j in range(col):
        val = (j+1)**(i+1)
        M[i].append(val)
M

[[1, 2, 3, 4, 5, 6, 7, 8],
 [1, 4, 9, 16, 25, 36, 49, 64],
 [1, 8, 27, 64, 125, 216, 343, 512],
 [1, 16, 81, 256, 625, 1296, 2401, 4096]]

In [197]:
n = len(M)
m = len(M[0])
r = 0
c = m-1
while r <= n-1 and c >= 0:
    print(M[r][c])
    c -=  1
    r += 1

8
49
216
625


## Given a rectangular matrix of size n*m , print all right to left diagonals. : 

#### Given a matrix of size 5x4, how many diagonals will be there?
    8


    5*4 martrix : diagonals :  right to left direction 
    
    1 2 3 4 
    2 3 4 5 
    3 4 5 6 
    4 5 6 7 
    5 6 7 8
    
    for m * n matrix : 
    
    Given a matrix of size NxM, how many diagonals will be there?
    N+M-1

In [198]:
def print_diagonals(M):
    row = len(M)
    col = len(M[0])
    for c in range(col):
        r = 0
        
        while r <= n-1 and c >= 0:
            print(M[r][c], end= " ")
            c -=  1
            r += 1
        print()

    for r in range(1,row):
        c = col-1
        
        while r <= n-1 and c >= 0:
            print(M[r][c], end= " ")
            c -=  1
            r += 1
        print()

        

In [199]:
M = [[0]*col for _ in range(row)]
for i in range(row):
    for j in range(col):
        M[i][j]= (j+1)**(i+1)
M

[[1, 2, 3, 4, 5, 6, 7, 8],
 [1, 4, 9, 16, 25, 36, 49, 64],
 [1, 8, 27, 64, 125, 216, 343, 512],
 [1, 16, 81, 256, 625, 1296, 2401, 4096]]

     [1,  2,  3,   4,   5,    6,    7,    8],
     [1,  4,  9,  16,  25,   36,   49,   64],
     [1,  8, 27,  64, 125,  216,  343,  512],
     [1, 16, 81, 256, 625, 1296, 2401, 4096]]


In [200]:
print_diagonals(M)

1 
2 1 
3 4 1 
4 9 8 1 
5 16 27 16 
6 25 64 81 
7 36 125 256 
8 49 216 625 
64 343 1296 
512 2401 
4096 


In [201]:

M = [[1,  2,  3,  4,  2], 
     [1,  4,  9,  6,  3], 
     [1,  8,  7,  4,  7], 
     [1,  1,  8,  6,  9]]

print_diagonals(M)


1 
2 1 
3 4 1 
4 9 8 1 
2 6 7 1 
3 4 8 
7 6 
9 


## Transpose the matrix : 

In [202]:

def transpose_matrix(M):
    r,c = len(M),len(M[0])
    
    M_t = [[0]*r for _ in range(c)]
    
    for col in range(c):
        for row in range(r):
            M_t[col][row] = M[row][col]
        
    
    return M_t


M = [[1,  2,  3,  4,  2], 
     [1,  4,  9,  6,  3], 
     [1,  8,  7,  4,  7], 
     [1,  1,  8,  6,  9]]
transpose_matrix(M)

[[1, 1, 1, 1], [2, 4, 8, 1], [3, 9, 7, 8], [4, 6, 4, 6], [2, 3, 7, 9]]

In [203]:
M = np.array([[1,  2,  3,  4,  2], 
     [1,  4,  9,  6,  3], 
     [1,  8,  7,  4,  7], 
     [1,  1,  8,  6,  9]])
M.T

<IPython.core.display.Javascript object>

array([[1, 1, 1, 1],
       [2, 4, 8, 1],
       [3, 9, 7, 8],
       [4, 6, 4, 6],
       [2, 3, 7, 9]])

In [204]:
M = [[1,2,3,4,2,1,8], 
     [1,4,9,6,3,2,7], 
     [1,8,7,4,7,3,6], 
     [1,1,8,6,9,4,5]]
transpose_matrix(M)

[[1, 1, 1, 1],
 [2, 4, 8, 1],
 [3, 9, 7, 8],
 [4, 6, 4, 6],
 [2, 3, 7, 9],
 [1, 2, 3, 4],
 [8, 7, 6, 5]]

In [205]:
M = [[1,2,3],
    [4,5,6],
     [7,8,9]]
r = len(M)
 

for i in range(r//2):
    for j in range(r):
        M[i][j],M[j][i] = M[j][i],M[i][j]
M

[[1, 4, 7], [2, 5, 6], [3, 8, 9]]

## rotate the n*m matrix 90 degree clockwise :

In [206]:
M = [[1,2,3,4,5,0],
    [6,7,8,9,10,0],
    [11,12,13,14,15,0],
    [16,17,18,19,20,0]]
M

[[1, 2, 3, 4, 5, 0],
 [6, 7, 8, 9, 10, 0],
 [11, 12, 13, 14, 15, 0],
 [16, 17, 18, 19, 20, 0]]

In [207]:
Mt = transpose_matrix(M)
Mt

[[1, 6, 11, 16],
 [2, 7, 12, 17],
 [3, 8, 13, 18],
 [4, 9, 14, 19],
 [5, 10, 15, 20],
 [0, 0, 0, 0]]

In [208]:
r = len(Mt)
c = len(Mt[0])
M_h = [[0]*c for _ in range(r)]
 
for row in range(r):
    for col in range(c):
        M_h[row][c-col-1] = Mt[row][col]
M_h

[[16, 11, 6, 1],
 [17, 12, 7, 2],
 [18, 13, 8, 3],
 [19, 14, 9, 4],
 [20, 15, 10, 5],
 [0, 0, 0, 0]]

In [209]:
def rotate90(M):
    def transpose_matrix(M):
        r,c = len(M),len(M[0])

        M_t = [[0]*r for _ in range(c)]

        for col in range(c):
            for row in range(r):
                M_t[col][row] = M[row][col]


        return M_t
    
    def horizontal_flip(Mt):
    
        r = len(Mt)
        c = len(Mt[0])
        M_h = [[0]*c for _ in range(r)]

        for row in range(r):
            for col in range(c):
                M_h[row][c-col-1] = Mt[row][col]
        return M_h
    
    return horizontal_flip(transpose_matrix(M))

In [210]:
M = [[18,25,43,44,45,18],
    [17,47,48,98,10,17],
    [11,12,13,14,15,16],
    [16,17,18,19,20,19]]
rotate90(M)

[[16, 11, 17, 18],
 [17, 12, 47, 25],
 [18, 13, 48, 43],
 [19, 14, 98, 44],
 [20, 15, 10, 45],
 [19, 16, 17, 18]]

## Matrix Multiplication : 

In [211]:
u = [[1,2,3],
     [4,5,6]] # 2*3    r*c

v = [[1,4,7,8],     # 3*4   c*p
     [2,5,8,2],
     [3,6,9,3]]
# output matrxi will be 2*4 r*p
r = len(u)
c  = len(u[0])
p = len(v[0])

ansMatrix = [[0]*p for _ in range(r)]
ansMatrix

[[0, 0, 0, 0], [0, 0, 0, 0]]

In [212]:
# ansMatrix(i,j) = sum of (firstMatrix(i row)) * (secondMatrix j col)

# ansMatrix(i,j) = sigma from k = 0 ( A(i,k) * B(k,j) )

def mat_dot_product(u,v):   # TC : r*c*p  and SC: r*p
    r = len(u)
    p = len(v[0])
    c = len(u[0])
    c = len(v)
    ansMatrix = [[0]*p for _ in range(r)]

    for i in range(r):
        for j in range(p):
            summs = 0
            for k in range(c):
                summs += u[i][k]*v[k][j]
            ansMatrix[i][j] = summs
            
    return ansMatrix

In [213]:
mat_dot_product(u,v)

[[14, 32, 50, 21], [32, 77, 122, 60]]

##  Add the matrices


#### You are given two matrices A & B of same size, you have to return another matrix which is the sum of A and B.


    Input Format
    First argument is vector of vector of integers representing matrix A.
    Second argument is vecotor of vector of integers representing matrix B.

    Output Format
    You have to return a vector of vector of integers after doing required operations.

    Example Input
    Input 1:
    A = [[1, 2, 3],[4, 5, 6],[7, 8, 9]]
    B = [[9, 8, 7],[6, 5, 4],[3, 2, 1]]

    Example Output
    Output 1:
    [[10, 10, 10], [10, 10, 10], [10, 10, 10]]

    Example Explanation
    Explanation 1:
    A + B = [[1+9, 2+8, 3+7],[4+6, 5+5, 6+4],[7+3, 8+2, 9+1]] = [[10, 10, 10], [10, 10, 10], [10, 10, 10]].

In [214]:
def matrix_addition(A,B):
    r = len(A)
    c = len(A[0])
    resulting_matrix = [[0]*c for _ in range(r)]

    for i in range(r):
        for j in range(c):
            resulting_matrix[i][j] = A[i][j] + B[i][j]
    return (resulting_matrix)

In [215]:
A = [[1, 2, 3],[4, 5, 6],[7, 8, 9]]
B = [[9, 8, 7],[6, 5, 4],[3, 2, 1]]
matrix_addition(A,B)

[[10, 10, 10], [10, 10, 10], [10, 10, 10]]

## Matrix Subtraction

In [216]:
def matrix_Subtraction(A,B):
    r = len(A)
    c = len(A[0])
    resulting_matrix = [[0]*c for _ in range(r)]

    for i in range(r):
        for j in range(c):
            resulting_matrix[i][j] = A[i][j] - B[i][j]
    return (resulting_matrix)



In [217]:
A = [[1, 2, 3],           
  [4, 5, 6],
  [7, 8, 9]]           
B = [[9, 8, 7],
  [6, 5, 4],
  [3, 2, 1]]
matrix_Subtraction(A,B)

[[-8, -6, -4], [-2, 0, 2], [4, 6, 8]]

## Column Sum
### You are given a 2D integer matrix A, return a 1D integer array containing column-wise sums of original matrix

In [218]:
def column_sum(A):

    n = len(A)
    m = len(A[0])
    col_sum = []
    for col in range(m):
        sum = 0
        for row in range(n):
            sum += A[row][col]
        col_sum.append(sum)
    return (col_sum)

column_sum([[1,2,3,4],
            [5,6,7,8],
            [9,2,3,4]])

[15, 10, 13, 16]

### You are given a 2D integer matrix A, return a 1D integer array containing row-wise sums of original matrix

In [219]:
def row_sum(A):

    n = len(A)
    m = len(A[0])
    row_sum = []
    for row in range(n):
        sum = 0
        for col in range(m):
            sum += A[row][col]
        row_sum.append(sum)
    return (row_sum)

row_sum([[1,2,3,4],
            [5,6,7,8],
            [9,2,3,4]])

[10, 26, 18]

## Row to Column Zero

### You are given a 2D integer matrix A, make all the elements in a row or column zero if the A[i][j] = 0. Specifically, make entire ith row and jth column zero.


    Input Format
    First argument is a vector of vector of integers.(2D matrix).
    Output Format
    Return a vector of vector after doing required operations.
    Example Input
    Input 1:
    [1,2,3,4]
    [5,6,7,0]
    [9,2,0,4]
    Example Output
    Output 1:
    [1,2,0,0]
    [0,0,0,0]
    [0,0,0,0]
    Example Explanation
    Explanation 1:
    A[2][4] = A[3][3] = 0, so make 2nd row, 3rd row, 3rd column and 4th column zero.

In [220]:
A = [[1,2,3,4],
[5,6,7,0],
[9,2,0,4]]

In [221]:
A

[[1, 2, 3, 4], [5, 6, 7, 0], [9, 2, 0, 4]]

In [222]:
r = len(A)
c = len(A[0])


In [223]:
for i in range(r):
    flag  = 0
    for j in range(c):
        if A[i][j] == 0:
            flag = 1
        if flag == 1:
            for j in range(c):
                if A[i][j] != 0:
                    A[i][j] = -1
    
for j in range(c):
    flag = 0
    for i in range(r):
        if A[i][j] == 0:
            flag = 1
    if flag == 1:
        for i in range(r):
            if A[i][j] != 0:
                A[i][j] = -1
                
for i in range(r):
    for j in range(c):
        if A[i][j] == -1:
            A[i][j] = 0
A

[[1, 2, 0, 0], [0, 0, 0, 0], [0, 0, 0, 0]]

In [224]:
A

[[1, 2, 0, 0], [0, 0, 0, 0], [0, 0, 0, 0]]

In [225]:
[[1, 2, 3, 4], 
[-1, -1, -1, 0], 
 [-1, -1, 0, -1]]

[[1, 2, 3, 4], [-1, -1, -1, 0], [-1, -1, 0, -1]]

### Are Matrices Same 

In [226]:
def Are_Matrices_Same(A,B):

        n = len(A)
        m = len(A[0])
        ans = True
        for i in range(n):
            for j in range(m):
                if A[i][j]!=B[i][j]:
                    
                    ans = False
        if ans:
            return 1
        else:
            return 0

In [227]:
A = [[1, 2, 3],[4, 5, 6],[7, 8, 9]]
B = [[1, 2, 3],[4, 5, 6],[7, 8, 9]]
Are_Matrices_Same(A,B)

1

## Print boundry elements of a given matrix in clockwise direction : 

In [228]:
r = 4
c = 8
M = [[0]*c for _ in range(r)]


In [229]:
x = 1
for j in range(c):
    for i in range(r):
        M[i][j] = x
        x = x+1
M

[[1, 5, 9, 13, 17, 21, 25, 29],
 [2, 6, 10, 14, 18, 22, 26, 30],
 [3, 7, 11, 15, 19, 23, 27, 31],
 [4, 8, 12, 16, 20, 24, 28, 32]]

In [230]:
def print_boundry(M):
    r = len(M)
    c = len(M[0])
    ans = []
    for j in range(c):
        ans.append(M[0][j])
    for i in range(1,r):
        ans.append(M[i][c-1])
    for j in reversed(range(c-1)):
        ans.append(M[r-1][j])
    for i in range(r-2,0,-1):
        ans.append(M[i][0])
    return ans

print_boundry(M)

[1, 5, 9, 13, 17, 21, 25, 29, 30, 31, 32, 28, 24, 20, 16, 12, 8, 4, 3, 2]

In [231]:
M

[[1, 5, 9, 13, 17, 21, 25, 29],
 [2, 6, 10, 14, 18, 22, 26, 30],
 [3, 7, 11, 15, 19, 23, 27, 31],
 [4, 8, 12, 16, 20, 24, 28, 32]]

In [232]:
def print_boundry(M):
    n = len(M)
    m = len(M[0])
    
    
    r,c = 0,0
    direction = "right"
    
    while True:
        
        print(M[r][c], end = " ")
        
        if direction == "right":
            if c == m-1:
                direction = "down"
                r += 1
            else:
                c+=1
                
        elif direction == "down":
            if r == n-1:
                direction = "left"
                c -= 1
            else:
                r += 1
    
        elif direction == "left":
            if c == 0:
                direction = "up"
                r -= 1
            else:
                c -= 1
        
        elif direction == "up":
            if r == 1:
                break
            else:
                r -= 1
                
        
    
        
        

print_boundry(M)

1 5 9 13 17 21 25 29 30 31 32 28 24 20 16 12 8 4 3 2 

In [233]:
M

[[1, 5, 9, 13, 17, 21, 25, 29],
 [2, 6, 10, 14, 18, 22, 26, 30],
 [3, 7, 11, 15, 19, 23, 27, 31],
 [4, 8, 12, 16, 20, 24, 28, 32]]

## Print spiral  : 

In [234]:
def print_boundry(M):
    n = len(M)
    m = len(M[0])
    
    
    r,c = 0,0
    direction = "right"
    boundary  = 0
    
    for _ in range(n*m):
        
        print(M[r][c], end = " ")
        
        if direction == "right":
            if c == m-1 - boundary:
                direction = "down"
                r += 1
            else:
                c+=1
                
        elif direction == "down":
            if r == n-1 - boundary:
                direction = "left"
                c -= 1
            else:
                r += 1
    
        elif direction == "left":
            if c == 0 + boundary:
                direction = "up"
                r -= 1
            else:
                c -= 1
        
        elif direction == "up":
            if r == 1 + boundary:
                direction = "right"
                boundary += 1
                c += 1
            else:
                r -= 1
                
        
    
        
        

print_boundry(M)

1 5 9 13 17 21 25 29 30 31 32 28 24 20 16 12 8 4 3 2 6 10 14 18 22 26 27 23 19 15 11 7 

In [235]:
M

[[1, 5, 9, 13, 17, 21, 25, 29],
 [2, 6, 10, 14, 18, 22, 26, 30],
 [3, 7, 11, 15, 19, 23, 27, 31],
 [4, 8, 12, 16, 20, 24, 28, 32]]

## Spiral Order Matrix II

### Given an integer A, generate a square matrix filled with elements from 1 to A2 in spiral order and return the generated square matrix.

    Input Format
    First and only argument is integer A
    Output Format
    Return a 2-D matrix which consists of the elements added in spiral order.
    Example Input
    Input 1:
    1
    Input 2:
    2
    Input 3:
    5
    Example Output
    Output 1:
    [ [1] ]
    Output 2:
    [ [1, 2],
      [4, 3] ]
    Output 2:

    [ [1,  2,  3,  4,  5], 
      [16, 17, 18, 19, 6], 
      [15, 24, 25, 20, 7], 
      [14, 23, 22, 21, 8], 
      [13, 12, 11, 10, 9] ]


    Example Explanation
    Explanation 1:

    Only 1 is to be arranged.
    Explanation 2:

    1 --> 2
          |
          |
    4<--- 3

In [236]:
def generate_spiral_matrix(A):
    
    M = [[0]*A for _ in range(A)]
    
    n = len(M)
    m = len(M[0])
    
    r , c = 0, 0
    direction = "right"
    boundary  = 0
    value = 1
    
    for _ in range(n*m):
        M[r][c] = value 
        
        if direction =="right":
            if c == m-1-boundary:
                direction = "down"
                r+=1
            else:
                c += 1
                
        elif direction == "down":
            
            if r == n-1 - boundary:
                direction = "left"
                c -= 1
            else:
                r += 1
                
        elif direction == "left":
            
            if c == 0 + boundary:
                direction = "up"
                r-=1
            else:
                c -= 1
        
        elif direction == "up":
            
            if r == 1+boundary:
                direction = "right"
                boundary += 1
                c += 1
            else:
                r -= 1
        value += 1
    return M

generate_spiral_matrix(5)           

[[1, 2, 3, 4, 5],
 [16, 17, 18, 19, 6],
 [15, 24, 25, 20, 7],
 [14, 23, 22, 21, 8],
 [13, 12, 11, 10, 9]]

## Spiral Order Matrix 2

### Given an integer A, Print an A X A square matrix filled with elements from 1 to A2 in spiral order.

    Input Format
    First and only line contains a single integer A.
    Output Format
    Print the spiral order pattern in A lines each containing A space-separated integers.
    Example Input
    Input 1:
     1 
    Input 2:
     2 
    Input 3:
     3 

    Example Output
    Output 1:
     1 
    Output 2:
     1 2 
     4 3 
    Output 3:
     1 2 3 
     6 5 4 
     7 8 9 



In [237]:
def spiral_order_matrix_2(A):
    M = [[0]*A for _ in range(A)]
    
    n = len(M)
    m = len(M[0])
    x = 1
    for i in range(n):
        for j in range(m):
            
            M[i][j] =  x
            x += 1
    
    for i in range(n):
        if i % 2 != 0:
            M[i] = M[i][::-1]
    
    return M
    
    
spiral_order_matrix_2(8)

[[1, 2, 3, 4, 5, 6, 7, 8],
 [16, 15, 14, 13, 12, 11, 10, 9],
 [17, 18, 19, 20, 21, 22, 23, 24],
 [32, 31, 30, 29, 28, 27, 26, 25],
 [33, 34, 35, 36, 37, 38, 39, 40],
 [48, 47, 46, 45, 44, 43, 42, 41],
 [49, 50, 51, 52, 53, 54, 55, 56],
 [64, 63, 62, 61, 60, 59, 58, 57]]

## Christmas Trees

### You are given an array A consisting of heights of Christmas trees and an array B of the same size consisting of the cost of each of the trees (Bi is the cost of tree Ai, where 1 ≤ i ≤ size(A)), and you are supposed to choose 3 trees (let's say, indices p, q, and r), such that Ap < Aq < Ar, where p < q < r.  The cost of these trees is Bp + Bq + Br. You are to choose 3 trees such that their total cost is minimum. Return that cost.  If it is not possible to choose 3 such trees return -1.

    Input Format
    First argument is an integer array A.
    Second argument is an integer array B.
    Output Format
    Return an integer denoting the minimum cost of choosing 3 trees whose heights are strictly in increasing order, if not possible, -1.
    Example Input
    Input 1:
     A = [1, 3, 5]
     B = [1, 2, 3]
    Input 2:
     A = [1, 6, 4, 2, 6, 9]
     B = [2, 5, 7, 3, 2, 7]
    Example Output
    Output 1:
     6 
    Output 2:
     7 
    Example Explanation
    Explanation 1:
     We can choose the trees with indices 1, 2 and 3, and the cost is 1 + 2 + 3 = 6. 
    Explanation 2:
     We can choose the trees with indices 1, 4 and 5, and the cost is 2 + 3 + 2 = 7. 
     This is the minimum cost that we can get.

In [238]:
H = [9,7,15,3,6,1,2,8 ,2,11,5,13,15,6]
C = [6,4, 1,5,8,5,9,11,6, 3,3, 2, 1,4]
ans = None
n = len(H)
mincost = float("inf")
for t1 in range(n):
    for t2 in range(t1,n):
        for t3 in range(t2,n):
            if H[t1] < H[t2] < H[t3]:
                cost = C[t1] + C[t2] + C[t3]
        
                if cost < mincost:
                    mincost = cost
#                     print("Heights are :", H[t1],H[t2],H[t3],"\ncosts are:",C[t1],C[t2],C[t3] , " mincost is :",mincost)
                    ans = [H[t1],H[t2],H[t3]]
print(ans)
print(mincost)
                
                

[11, 13, 15]
6


In [239]:
H = [9,7,15,3,6,1,2,8 ,2,11,5,13,15,6]
C = [6,4, 1,5,8,5,9,11,6, 3,3, 2, 1,4]
ans = None
n = len(H)
mincost = float("inf")

for t2 in range(n):
    
    Ht2 = H[t2]  # height of tree2
    
    Ct2 = C[t2]   # cost of tree2
    
    Ct1 = float("inf")
    
    for t1 in range(t2):    # finding minimum cost tree1 having height < t2 height
        if H[t1] < H[t2]:
            Ct1 = min(C[t1],Ct1)
            
    Ct3 = float("inf")
    
    for t3 in range(t2+1,n):  # finding minimum cost tree3 having height > t2 height
        if H[t3] > H[t2]:
            Ct3 = min(C[t3],Ct3)
            
        
    cost = Ct1 + Ct2 + Ct3
    mincost = min(cost,mincost)
    
print(mincost)
                  

6


## Maximum positivity

### Given an array of integers A, of size N.  Return the maximum size subarray of A having only non-negative elements. If there are more than one such subarray, return the one having the smallest starting index in A.

    Input Format
    The first and only argument given is the integer array A.
    Output Format
    Return maximum size subarray of A having only non-negative elements. If there are more than one such subarrays, return the one having earliest starting index in A.
    Example Input
    Input 1:
     A = [5, 6, -1, 7, 8]
    Input 2:
     A = [1, 2, 3, 4, 5, 6]
    Example Output
    Output 1:
     [5, 6]
    Output 2:
     [1, 2, 3, 4, 5, 6]
    Example Explanation
    Explanation 1:
     There are two subarrays of size 2 having only non-negative elements.
     1. [5, 6]  starting point  = 0
     2. [7, 8]  starting point  = 3
     As starting point of 1 is smaller, return [5, 6]
    Explanation 2:
     There is only one subarray of size 6 having only non-negative elements:
     [1, 2, 3, 4, 5, 6]


In [240]:
A = [5,6,-1,7,8]
def Maximum_positivity(A):
    n = len(A)
    ans  = None
    maxsize = -float("inf")
    for s in range(n):
        for e in range(s,n):
            hasneg = False
            for i in range(s,e+1):
                if A[i] < 0: 
                    hasneg = True
                    break
            l = len(A[s:e+1])
            if l > maxsize:
                maxsize = l

                if hasneg == False:
                    ans = (A[s:e+1])
            
    return (ans)
Maximum_positivity(A)


[5, 6]

In [241]:

Maximum_positivity([1, 2, 3, 4, 5, 6])


[1, 2, 3, 4, 5, 6]

In [242]:
print(Maximum_positivity([ -6148706, -8550681, 4494391, 8074193, -1283236, 4384884, 8111626, -8972065, 7428432, 7514578 ]))

None


In [243]:

def Maximum_positivity(A):
#     print(A)
    ret = [[]]
    for i in A:
        if i < 0:
            ret.append([])
        else:
            ret[-1].append(i)
#     print(ret)
    mxlen = max([len(i) for i in ret])
    for i in ret:
        if len(i) == mxlen:
            return i
Maximum_positivity(A)

[5, 6]

In [244]:
print(Maximum_positivity([ -6148706, -8550681, 4494391, 8074193, -1283236, 4384884, 8111626, -8972065, 7428432, 7514578 ]))

[4494391, 8074193]


## Convert from binary to decimal : 

In [245]:
S = '1001110'

In [246]:
def BtoD(S):
    ans = 0
    x = 0
    for i in S[::-1]:
        ans += int(i)*(2**x)
        x+= 1
    return ans

In [247]:
BtoD(S)

78

### Convert given ternary to decimal ?
(02101)


In [248]:
S = '222020'

ans = 0
x = len(S)-1
for i in S:
    ans += int(i)*(3**x)
    x -= 1
ans

708

## Decimal to Binary : 

In [249]:
def DtoB(N):
    
    string = ""
    
    while N > 0:
        digit = N%2 
        string = str(digit) + string
        N = N // 2
    return string
        
    

In [250]:
DtoB(78),bin(78)[2:]

('1001110', '1001110')

In [251]:
def DtoT(N):
    
    string = ""
    
    while N > 0:
        digit = N%3
        string = str(digit) + string
        N = N // 3
    return string
        
DtoT(708)

'222020'

## Quick way to convert Binary to Hexadecimal  : 

In [252]:
S = "10110111"

In [253]:
n = len(S)

In [254]:
S[:n//2],S[n//2:]

('1011', '0111')

In [255]:
BtoD(S[:n//2]),BtoD(S[n//2:])


(11, 7)

In [256]:
# in Hexadecimal 11 is B 
ans = 'B7'

# `Bitwise operators : `

# ans operator : &


In [257]:
12 & 2

0

In [258]:
bin(12)[2:]

'1100'

In [259]:
bin(2)[2:]

'10'

        1100
      & 0010

        0000

In [260]:
bin(10)[2:],bin(2)[2:]

('1010', '10')

        1010
       &0010
       
        0010

In [261]:
BtoD("10")

2

In [262]:
a = 13
b = 10
a&b 

8

In [263]:
bin(13)[2:],bin(10)[2:]   # 13 & 10

('1101', '1010')

1101
1010

1000



In [264]:
BtoD("1000")

8

In [265]:
# usecase of and operator if a bit is on or off : 



### int a = x // random value x is assigned to  print(a&1)
    1 if a is odd
    0 if a is even
    

In [266]:
a = 15
a&1

1

In [267]:
a = 16
a&1

0

# OR ooperator: 


In [268]:

10|4

14

In [269]:
bin(10)[2:],bin(4)[2:] # 10 | 4

('1010', '100')

    1010
    0100
    
    1110
    

In [270]:
BtoD("1110")

14

In [271]:
bin(13)[2:],bin(10)[2:] # 13 | 10

('1101', '1010')

    1101
    1010

    1111

In [272]:
BtoD("1111")

15

### int a = x // random value x is assigned to a print(a|1)
    if a is odd x
    If a is even x+1

In [273]:
a = 15
a|1

15

In [274]:
a = 16
a|1

17

In [275]:
# use case of OR operator is to turn on the bit


# XOR operator ^ 

- usecase : toggle the bit


In [276]:
bin(13)[2:],bin(14)[2:]

('1101', '1110')

    1101
    0001
    
    1100
    
    
    1110
    0001
    
    1111
    
    

In [277]:
13^1,14^1

(12, 15)

In [278]:
16^1

17

In [279]:
14^1

15

    0 0 > 0 
    1 0 > 1
    0 1 > 1
    1 1 > 0

In [280]:
bin(13)[2:],bin(10)[2:] # 13 ^ 10

('1101', '1010')

    1101
    1010
    
    0111
    
    

In [281]:
BtoD("0111")

7

In [282]:
13^10

7

### int a = x : random value x is assigned to a print(a^1)
    if a is even = x+1
    if a is odd = x-1

## Single Number

### Given an array of integers A, every element appears twice except for one. Find that integer that occurs once.

    Input Format
    The first and only argument of input contains an integer array A.
    Output Format
    Return a single integer denoting the single element.

    Example Input
    Input 1:
     A = [1, 2, 2, 3, 1]
    Input 2:
     A = [1, 2, 2]
    Example Output
    Output 1:
     3
    Output 2:
     1

    Example Explanation
    Explanation 1:
    3 occurs once.
    Explanation 2:
    1 occurs once.

In [283]:
13^13

0

In [284]:
5^5^3^19^3

19

In [285]:
A = [5,5,2,1,2,3,9,9,3,1,0,-1,-1,8]
ans = A[0]
for i in range(1,len(A)):
    ans = ans ^ A[i]
ans

8

In [286]:
A = [5,5,2,1,2,3,9,9,3,1,0,-1,-1,8]
ans = 0
for i in range(len(A)):
    ans = ans ^ A[i]
ans

8

### # check if the 1st bit is on or off ! :


In [287]:

DtoB(44)

'101100'

    101100
    000001

    000000


In [288]:
44&1

0

### # check if the 2st bit is on or off ! :


In [289]:
DtoB(44)

'101100'

    101100
    000010

    000000

In [290]:
44&2

0

### # check if the 3rd bit is on or off ! :


    101100
    000100
    
    000100

In [291]:
44&4

4

In [292]:
44&8

8

In [293]:
44&16

0

In [294]:
44&32

32

In [295]:
44&64

0

In [296]:
44&128

0

In [297]:
[44 & (2**0),
 44 & (2**1),
 44 & (2**2),
 44 & (2**3),
 44 & (2**4),
 44 & (2**5),
]

[0, 0, 4, 8, 0, 32]

In [298]:
[44 & (2**0),
 44 & (2**1),
 44 & (2**2),
 44 & (2**3),
 44 & (2**4),
 44 & (2**5),
][::-1]

[32, 0, 8, 4, 0, 0]

In [299]:
101100

101100

## check if kth bit is on or off : 

    Number & 2**(k-1)
    

In [300]:
bin(998)[2:]

'1111100110'

In [301]:
# checking if 8th bit in 998 is on or off : 

In [302]:
998&(2**7)

128

In [303]:
998&(2**4)  # checking  5th bit is on or off : 

0

In [304]:
[bin(11)[2:],
 bin(22)[2:],
 bin(44)[2:],
 bin(88)[2:],
 bin(176)[2:],
]

['1011', '10110', '101100', '1011000', '10110000']

    shifting all bits towards left 

In [305]:
11 << 1

22

In [306]:
11 << 2

44

In [307]:
11 << 3

88

In [308]:
11 << 4

176

In [309]:
[10<<1,10<<2,10<<3,10<<4,10<<5]

[20, 40, 80, 160, 320]

In [310]:
[DtoB(1<<1),
 DtoB(1<<2),
 DtoB(1<<3),
 DtoB(1<<4),
 DtoB(1<<5),
 DtoB(1<<6),
 DtoB(1<<7),
 DtoB(1<<8)
 
]

['10', '100', '1000', '10000', '100000', '1000000', '10000000', '100000000']

    (a << x)  is equal to  (a * 2^x)

## checking ith bit is on or off : 

In [311]:
bin(44)[2:]

'101100'

In [312]:
[44 & (2**0),
 44 & (2**1),
 44 & (2**2),
 44 & (2**3),
 44 & (2**4),
 44 & (2**5),
]

[0, 0, 4, 8, 0, 32]

In [313]:
[44 & (1 <<0),
 44 & (1 <<1),
 44 & (1 <<2),
 44 & (1 <<3),
 44 & (1 <<4),
 44 & (1 <<5)]

[0, 0, 4, 8, 0, 32]

In [314]:
def check_ith_bit_on_or_off(n,i):
    if i <= 0: return 'check index entered negative or 0'
    if n & (1<<(i-1)) == 0:
        return False
    return True
    
    

In [315]:
check_ith_bit_on_or_off(44,7)

False

In [316]:
check_ith_bit_on_or_off(7,4),check_ith_bit_on_or_off(7,3),check_ith_bit_on_or_off(7,1),check_ith_bit_on_or_off(7,0)

(False, True, True, 'check index entered negative or 0')

In [317]:
bin(7)[2:]

'111'

In [318]:
32>>1,32>>2,32>>4

(16, 8, 2)

### Which of the following options output is 2 power k
    
    1 << k

### check 422 , 5th bit on or off 

In [319]:
BtoD("110100110")

422

In [320]:
422 & (1<<(6-1)) # 6th bit on or off

32

In [321]:
422 & (1<<(8-1)) # 8th bit on or off

128

In [322]:
422 & (1<<(4-1)) # 4th bit on or off

0

## count the number of set bits : 

In [323]:
N = 422

def count_set_bits(N):
    count = 0
    k = 0
    while True :
        if N & (1<<(k)) != 0:
            count += 1
        k += 1
        if N < (1<<k):
            break
    return count
            
count_set_bits(N)      

5

In [324]:
DtoB(422)

'110100110'

In [325]:
N = 422

def count_set_bits(N):
    count = 0
    
    for k in range(int(math.log(N,2))+2):
        
        if N & (1<<(k)) != 0:
            count += 1
        
    return count
            
count_set_bits(N)      

5

In [326]:
len(bin(N)[2:])

9

In [327]:
int(math.log(422,2))+1

9

In [328]:
N = 422

def count_set_bits(N):
    count = 0
    
    while N > 0:
        count += N&1
        print(N, " | ",N&1," | " ,count )
        N = N>>1
        
    return count
count_set_bits(N)      

422  |  0  |  0
211  |  1  |  1
105  |  1  |  2
52  |  0  |  2
26  |  0  |  2
13  |  1  |  3
6  |  0  |  3
3  |  1  |  4
1  |  1  |  5


5

## How negative numbers represented in binary  :

    2's compliment : 
        1: find binary of magnitud
        2: toggle all bits
        3: add 1

In [329]:
-40

-40

In [330]:
bin(40)[2:]

'101000'

            00101000    in 8 bit format : 
     ---------------------
     toggle all bits now : 
            11010111
      --------------------------------
      add 1 
      --------------------------------
            11010111
                   1
                   
                   1+1  = 10
              ------------------ 
                111 
            11010111
            +      1
            11011000
            
            



In [331]:

BtoD("10000000"),BtoD("1011000")

(128, 88)

In [332]:
-128 + 88 

-40

In [333]:
# toggle all the bits

In [334]:
40

40

In [335]:
bin(40^0)[2:]

'101000'

In [336]:
bin(40^1)[2:]

'101001'

In [337]:
bin(40^2)[2:]

'101010'

In [338]:
bin(40^4)[2:]

'101100'

In [339]:
bin(40^8)[2:]

'100000'

In [340]:
bin(40^16)[2:]

'111000'

In [341]:
bin(40^32)[2:]

'1000'

In [342]:

# Python3 program to toggle set
# bits starting from MSB
from math import log
 
# Returns a number which has all set bits
# starting from MSB of n
def toggle(num):
 
    # the number of bits is equal to log2num + 1
    n = int(log(num, 2)) + 1
 
    # calculating mask
    mask = pow(2, n) - 1
 
    # toggling bits using xor with mask
    return num ^ mask
 
# Driver code
num = 10
print(bin(num)[2:])
print(bin(toggle(num))[2:])


1010
101


### -1 in binary : 

    -1 : 
    
    00000001
    
    toggleL:
    11111110
          +1
          
           1
    
    
    

In [343]:
BtoD("1111111")

127

In [344]:
BtoD("10000000")

128

In [345]:
127-128

-1

In [346]:
# 1   1  1  1 1 1 1 1
-128+64+32+16+8+4+2+1

-1

# Modular Arithmatic : 

In [347]:
100%7

2

In [348]:
(100-2)/7

14.0

In [349]:
100 == (14*7)+2

True

In [350]:
150%11

7

In [351]:
150==(11*13)+7

True

    (a+b)%k = ((a%k)+(b%k))%k
    
    (a*b)%k = ((a%k)*(b%k))%k

    (a^b)% M
    
    
    
    

In [352]:
a = 5
b = 100000
M = 97


In [353]:
ans = 1
for i in range(b):
    ans = ((ans % M)*(a % M)) % M
ans

61

In [354]:
(a**b)%M

61

     (a^b)%M = ((a%M)*(a^b-1%M))%M

###  Excel Column Number

### Given a column title as appears in an Excel sheet, return its corresponding column number.

    Problem Constraints
    1 <= length of the column title <= 5

    Input Format
    The only argument is a string that represents the column title in the excel sheet.

    Output Format
    Return a single integer that represents the corresponding column number.

    Example Input
    Input 1:
     AB
    Input 2:
     BB

    Example Output
    Output 1:
     28
    Output 2:
     54

    Example Explanation
    Explanation 1:

     A -> 1
     B -> 2
     C -> 3
     ...
     Z -> 26
     AA -> 27
     AB -> 28
    Explanation 2:

     A -> 1
     B -> 2
     C -> 3
     ...
     Z -> 26
     AA -> 27
     AB -> 28
     ...
     AZ -> 52
     BA -> 53
     BB -> 54


In [355]:
def titleTONumber(A):
    ans = 0
    for i in range(len(A)):
        ans *= 26
        ans += ord(A[i]) - ord("A") + 1
        print(ans)
    return ans

In [360]:
titleTONumber("BD")

2
56


56

 ### A, B and Modulo
 
 ### Given two integers A and B, find the greatest possible positive integer M,
 ### such that A % M = B % M.



       there possibilities 
       A > B
       B < A
       A = B
       
       LETS TAKE ONE : 
       A > B : 
           
               B = B
               B = A-(A-B)
               B%(A-B) = (A-(A-B))%(A-B)
               B%(A-B) = ( (A%(A-B)) - ((A-B)%(A-B)) )%(A-B)
               B%(A-B) = (A%(A-B))%(A-B) 
               B%(A-B) = A%(A-B)
               
               so ! 
               M = A-B
       
      
       

In [362]:
def AandBModulo(A, B):
    return abs(A-B)

### Leap year? - III

### Given an integer A representing a year, Return 1 if it is a leap year else, return 0. A year is a leap year if the following conditions are satisfied: The year is multiple of 400. Else the year is multiple of 4 and not multiple of 100.

    Input 1

     A = 2020
    Input 2:

     A = 1999


    Example Output
    Output 1

     1
    Output 2:

     0


    Example Explanation
    Explanation 1

     2020 is a leap year.
    Explanation 2:

     1999 is not a leap year.



In [367]:
def leapYear(year):
    if year % 400 == 0:
        return "LearYear"
    elif year % 4 ==0 and year % 100 != 0:
        return "LeapYear"
    return "Not a LeapYear"

In [368]:
leapYear(2020)

'LeapYear'

In [369]:
leapYear(1991)

'Not a LeapYear'

## LCM 

In [375]:
A = 18
B = 12
G = 1
for i in range(1,min(A,B)+1):
    if A % i == 0 and B % i == 0:
        G = i
    L = (A*B)/G
print(L)
        

36.0


In [376]:
(18*12)/6

36.0

In [ ]:
a